// strray-ai/dist/cli/commands/storyteller.js
/**
 * Storyteller CLI Command
 *
 * Invokes the storyteller skill to write reflections, sagas, journeys, or narratives.
 * Automatically determines story type based on context if not specified.
 *
 * Usage:
 *   npx strray-ai storyteller reflection
 *   npx strray-ai storyteller saga "v1.18.0 Release Journey"
 *   npx strray-ai storyteller journey --title "API Migration"
 *   npx strray-ai storyteller narrative --framework hero_journey
 */
import { readFileSync, existsSync, writeFileSync, mkdirSync } from "fs";
import { join } from "path";
import { execSync } from "child_process";
const STORY_TYPES = ["reflection", "saga", "journey", "narrative"];
function loadConfig(cwd) {
    const candidates = [
        join(cwd, ".opencode", "strray", "features.json"),
        join(cwd, ".strray", "features.json"),
    ];
    for (const path of candidates) {
        if (existsSync(path)) {
            try {
                const content = JSON.parse(readFileSync(path, "utf-8"));
                if (content.storytelling) {
                    return content.storytelling;
                }
            }
            catch {
                // ignore
            }
        }
    }
    return null;
}
function getStoryTypeMeta(storyType, config) {
    const defaults = {
        reflection: {
            location: "docs/reflections/",
            min_words: 2000,
            ideal_words: 5000,
            framework: "three_act_structure",
        },
        saga: {
            location: "docs/reflections/deep/",
            min_words: 5000,
            ideal_words: 15000,
            framework: "hero_journey",
        },
        journey: {
            location: "docs/reflections/deep/",
            min_words: 1500,
            ideal_words: 4000,
            framework: "three_act_structure",
        },
        narrative: {
            location: "docs/reflections/",
            min_words: 1000,
            ideal_words: 3000,
            framework: "three_act_structure",
        },
    };
    const configType = config?.story_types?.[storyType];
    return configType ? { ...defaults[storyType], ...configType } : defaults[storyType];
}
function getCommitCount() {
    try {
        const result = execSync("git rev-list --count HEAD 2>/dev/null", {
            encoding: "utf-8",
        });
        return parseInt(result.trim(), 10);
    }
    catch {
        return 0;
    }
}
function getRecentCommits(limit = 20) {
    try {
        return execSync(`git log --oneline -${limit} 2>/dev/null`, { encoding: "utf-8" });
    }
    catch {
        return "No git history available";
    }
}
function getChangedFiles() {
    try {
        return execSync("git diff --name-only HEAD~10 2>/dev/null || git ls-files 2>/dev/null | head -50", {
            encoding: "utf-8",
        });
    }
    catch {
        return "Unable to determine changed files";
    }
}
function generateFrontmatter(storyType, title, meta) {
    const date = new Date().toISOString().split("T")[0];
    const slug = title
        .toLowerCase()
        .replace(/[^a-z0-9]+/g, "-")
        .replace(/^-|-$/g, "");
    return `---
story_type: ${storyType}
title: "${title}"
date: ${date}
slug: ${slug}
emotional_arc: "challenge → discovery → resolution"
codex_terms: ["error-prevention", "systematic-validation"]
framework: ${meta.framework}
target_words: ${meta.ideal_words}
location: ${meta.location}
---
`;
}
function generatePrompt(storyType, title, meta, context, config) {
    const quality = config?.quality_requirements;
    return `Write a ${storyType} titled "${title}".

## Requirements
- **Framework**: ${meta.framework}
- **Target length**: ${meta.ideal_words} words (minimum: ${meta.min_words})
- **Location**: ${meta.location}

${quality?.require_frontmatter ? "- Include frontmatter with story_type, emotional_arc, codex_terms" : ""}
${quality?.require_key_takeaways ? "- End with a Key Takeaways section" : ""}
${quality?.require_what_next ? "- End with a What Next section" : ""}
${quality?.fact_check_before_publish ? "- Fact-check all technical details (agent names, file paths, error messages)" : ""}
${quality?.peer_review_agent ? "- Note: Send to @${quality.peer_review_agent} for peer review before publishing" : ""}

## Context
${context}

## Voice Guidelines
- Warmly candid and conversational
- Technical accuracy with narrative flow
- Include specific details: file paths, error messages, agent names
- Show the journey: challenges, discoveries, decisions

Please write the complete ${storyType} following these guidelines.
`;
}
function getOutputPath(storyType, title, meta) {
    const date = new Date().toISOString().split("T")[0];
    const slug = title
        .toLowerCase()
        .replace(/[^a-z0-9]+/g, "-")
        .replace(/^-|-$/g, "");
    const filename = `${storyType}-${slug}-${date}.md`;
    return join(meta.location, filename);
}
export async function storytellerCommand(storyType, options) {
    const cwd = process.cwd();
    const config = loadConfig(cwd);
    if (config && !config.enabled) {
        console.log("⚠️ Storytelling is disabled in features.json");
        return;
    }
    if (!storyType || !STORY_TYPES.includes(storyType)) {
        console.log("📖 0xRay Storyteller");
        console.log("========================");
        console.log("");
        console.log("Usage: npx strray-ai storyteller <type> [title]");
        console.log("");
        console.log("Story types:");
        console.log("  reflection  - Technical deep reflection (2-5k words)");
        console.log("  saga        - Long-form technical saga (5-15k words)");
        console.log("  journey     - Investigation/learning journey (1.5-4k words)");
        console.log("  narrative   - Technical narrative (1-3k words)");
        console.log("");
        console.log("Examples:");
        console.log('  npx strray-ai storyteller reflection "API Refactoring"');
        console.log('  npx strray-ai storyteller saga "v1.18.0 Release Journey"');
        console.log('  npx strray-ai storyteller journey --title "Bug Investigation"');
        console.log('  npx strray-ai storyteller narrative --framework hero_journey');
        return;
    }
    const type = storyType;
    const meta = getStoryTypeMeta(type, config);
    const title = options?.title || `${type.charAt(0).toUpperCase() + type.slice(1)} - ${new Date().toISOString().split("T")[0]}`;
    if (options?.framework) {
        meta.framework = options.framework;
    }
    const commitCount = getCommitCount();
    const recentCommits = getRecentCommits(20);
    const changedFiles = getChangedFiles();
    const context = `Recent commits (${commitCount} total):
${recentCommits}

Recent file changes:
${changedFiles}
`;
    const prompt = generatePrompt(type, title, meta, context, config);
    const outputPath = options?.output || getOutputPath(type, title, meta);
    console.log("📖 0xRay Storyteller");
    console.log("========================");
    console.log("");
    console.log(`Type: ${type}`);
    console.log(`Title: ${title}`);
    console.log(`Framework: ${meta.framework}`);
    console.log(`Target: ${meta.ideal_words} words (min: ${meta.min_words})`);
    console.log(`Output: ${outputPath}`);
    console.log("");
    if (options?.dryRun) {
        console.log("(Dry run - showing prompt that would be used)");
        console.log("");
        console.log("--- PROMPT ---");
        console.log(prompt);
        console.log("--- END PROMPT ---");
        return;
    }
    const frontmatter = generateFrontmatter(type, title, meta);
    const outputDir = join(cwd, meta.location);
    if (!existsSync(outputDir)) {
        mkdirSync(outputDir, { recursive: true });
    }
    const fullPath = join(cwd, outputPath);
    writeFileSync(fullPath, frontmatter + "\n" + prompt + "\n", "utf-8");
    console.log(`✅ Story template created: ${outputPath}`);
    console.log("");
    console.log("Next steps:");
    console.log("1. Open the file and write the story using the prompt as guidance");
    console.log("2. Follow the quality checklist in storyteller.yml");
    if (config?.quality_requirements?.peer_review_agent) {
        console.log(`3. Send to @${config.quality_requirements.peer_review_agent} for peer review`);
    }
    console.log("");
    console.log("💡 Tip: You can also ask the AI in your current session:");
    console.log(`   "Write a ${type} about ${title}"`);
}
export default storytellerCommand;
