// strray-ai/dist/delegation/agent-delegator.js
/**
 * Agent Delegator
 *
 * Intelligent agent delegation system that uses complexity analysis to determine
 * optimal task distribution strategies and conflict resolution.
 *
 * Integrates with TaskSkillRouter for keyword-based preprocessing.
 *
 * @version 1.1.0
 * @since 2026-01-07
 */
import { ComplexityAnalyzer, } from "./complexity-analyzer.js";
import { frameworkLogger } from "../core/framework-logger.js";
import { createTaskSkillRouter } from "./task-skill-router.js";
import { getKernel } from "../core/kernel-patterns.js";
/**
 * AgentDelegator class for intelligent task delegation
 */
export class AgentDelegator {
    complexityAnalyzer;
    stateManager;
    configLoader;
    taskSkillRouter;
    kernel;
    constructor(stateManager, configLoader) {
        this.stateManager = stateManager;
        this.configLoader = configLoader;
        this.complexityAnalyzer = new ComplexityAnalyzer();
        this.taskSkillRouter = createTaskSkillRouter(stateManager);
        this.kernel = getKernel();
    }
    getAvailableAgents() {
        const defaultAgents = [
            {
                name: "enforcer",
                capabilities: ["code-quality", "validation"],
                status: "active",
                expertise: "code quality enforcement",
                capacity: 100,
                performance: 95,
                specialties: ["validation", "compliance"],
            },
            {
                name: "architect",
                capabilities: ["design", "planning"],
                status: "active",
                expertise: "system architecture",
                capacity: 90,
                performance: 85,
                specialties: ["design", "planning"],
            },
            {
                name: "orchestrator",
                capabilities: [
                    "task-coordination",
                    "multi-agent-management",
                    "workflow-orchestration",
                ],
                status: "active",
                expertise: "orchestrating complex multi-agent workflows",
                capacity: 90,
                performance: 85,
                specialties: ["coordination", "delegation", "workflow"],
            },
            {
                name: "log-monitor",
                capabilities: ["log-analysis", "pattern-detection", "alerting"],
                status: "active",
                expertise: "log monitoring and analysis",
                capacity: 60,
                performance: 75,
                specialties: ["logs", "monitoring", "alerts"],
            },
            {
                name: "researcher",
                capabilities: [
                    "code-search",
                    "documentation-lookup",
                    "implementation-search",
                ],
                status: "active",
                expertise: "codebase search and documentation",
                capacity: 80,
                performance: 85,
                specialties: ["search", "docs", "patterns"],
            },
            {
                name: "multimodal-looker",
                capabilities: [
                    "image-analysis",
                    "diagram-understanding",
                    "visual-inspection",
                ],
                status: "active",
                expertise: "analyzing images and visual content",
                capacity: 70,
                performance: 80,
                specialties: ["images", "diagrams", "screenshots"],
            },
            {
                name: "code-analyzer",
                capabilities: ["code-analysis", "pattern-detection", "metrics"],
                status: "active",
                expertise: "deep code analysis",
                capacity: 75,
                performance: 80,
                specialties: ["analysis", "metrics", "patterns"],
            },
            {
                name: "code-reviewer",
                capabilities: ["review", "quality"],
                status: "active",
                expertise: "code review",
                capacity: 80,
                performance: 80,
                specialties: ["review", "quality"],
            },
            {
                name: "security-auditor",
                capabilities: ["security", "audit"],
                status: "active",
                expertise: "security analysis",
                capacity: 70,
                performance: 75,
                specialties: ["security", "audit"],
            },
            {
                name: "testing-lead",
                capabilities: ["testing", "coverage"],
                status: "active",
                expertise: "test architecture",
                capacity: 85,
                performance: 80,
                specialties: ["testing", "coverage"],
            },
            {
                name: "refactorer",
                capabilities: ["refactoring", "optimization"],
                status: "active",
                expertise: "code refactoring",
                capacity: 75,
                performance: 85,
                specialties: ["refactoring", "optimization"],
            },
            {
                name: "bug-triage-specialist",
                capabilities: ["debugging", "analysis"],
                status: "active",
                expertise: "bug triage",
                capacity: 60,
                performance: 70,
                specialties: ["debugging", "analysis"],
            },
            {
                name: "strategist",
                capabilities: [
                    "strategic-planning",
                    "complex-problem-solving",
                    "architecture-design",
                    "technical-strategy",
                    "risk-assessment",
                ],
                status: "active",
                expertise: "strategic guidance and complex problem-solving",
                capacity: 100,
                performance: 95,
                specialties: ["architecture decisions", "technical strategy", "risk analysis"],
            },
            {
                name: "seo-consultant",
                capabilities: [
                    "technical-seo-audit",
                    "schema-markup-generation",
                    "robots-txt-optimization",
                    "core-web-vitals-optimization",
                    "ai-search-optimization",
                ],
                status: "active",
                expertise: "technical SEO optimization",
                capacity: 70,
                performance: 80,
                specialties: ["schema", "robots.txt", "Core Web Vitals", "AI search"],
            },
            {
                name: "content-creator",
                capabilities: [
                    "seo-content-writing",
                    "keyword-optimization",
                    "meta-description",
                    "content-strategy",
                ],
                status: "active",
                expertise: "SEO content creation",
                capacity: 65,
                performance: 75,
                specialties: ["content", "keywords", "metadata"],
            },
            {
                name: "growth-strategist",
                capabilities: [
                    "campaign-strategy",
                    "market-analysis",
                    "brand-positioning",
                    "content-marketing-strategy",
                ],
                status: "active",
                expertise: "strategic marketing",
                capacity: 75,
                performance: 80,
                specialties: ["campaigns", "branding", "growth"],
            },
            {
                name: "database-engineer",
                capabilities: [
                    "schema-design",
                    "query-optimization",
                    "migrations",
                    "data-modeling",
                ],
                status: "active",
                expertise: "database architecture",
                capacity: 70,
                performance: 75,
                specialties: ["schema", "performance", "migrations"],
            },
            {
                name: "devops-engineer",
                capabilities: [
                    "ci-cd",
                    "infrastructure",
                    "containerization",
                    "monitoring",
                ],
                status: "active",
                expertise: "DevOps and infrastructure",
                capacity: 70,
                performance: 75,
                specialties: ["CI/CD", "Kubernetes", "AWS"],
            },
            {
                name: "backend-engineer",
                capabilities: [
                    "api-design",
                    "server-logic",
                    "database-integration",
                    "security",
                ],
                status: "active",
                expertise: "backend development",
                capacity: 80,
                performance: 85,
                specialties: ["APIs", "Node.js", "databases"],
            },
            {
                name: "frontend-engineer",
                capabilities: [
                    "ui-development",
                    "component-design",
                    "state-management",
                    "accessibility",
                ],
                status: "active",
                expertise: "frontend development",
                capacity: 80,
                performance: 85,
                specialties: ["React", "TypeScript", "CSS"],
            },
            {
                name: "performance-engineer",
                capabilities: [
                    "performance-optimization",
                    "profiling",
                    "caching",
                    "load-testing",
                ],
                status: "active",
                expertise: "performance optimization",
                capacity: 65,
                performance: 80,
                specialties: ["profiling", "caching", "optimization"],
            },
            {
                name: "mobile-developer",
                capabilities: [
                    "ios-development",
                    "android-development",
                    "cross-platform",
                    "mobile-ui",
                ],
                status: "active",
                expertise: "mobile app development",
                capacity: 70,
                performance: 75,
                specialties: ["React Native", "iOS", "Android"],
            },
            {
                name: "tech-writer",
                capabilities: [
                    "api-documentation",
                    "markdown",
                    "technical-writing",
                    "examples",
                ],
                status: "active",
                expertise: "technical documentation",
                capacity: 70,
                performance: 80,
                specialties: ["API docs", "guides", "READMEs"],
            },
        ];
        return defaultAgents.map((agent) => {
            const storedAgent = this.stateManager.get(`agent_capabilities:${agent.name}`);
            return storedAgent || agent;
        });
    }
    /**
     * Pre-process a task description using TaskSkillRouter
     * This extracts operation type and context from natural language descriptions
     * before running complexity analysis
     */
    preprocessTaskDescription(description, options) {
        const routingOptions = {
            ...(options?.sessionId && { sessionId: options.sessionId }),
            ...(options?.taskId && { taskId: options.taskId }),
            ...(options?.complexity !== undefined && {
                complexity: options.complexity,
            }),
            stateManager: this.stateManager,
        };
        const preprocessResult = this.taskSkillRouter.preprocess(description, routingOptions);
        return {
            operation: preprocessResult.operation,
            context: preprocessResult.context,
            suggestedAgent: preprocessResult.routing.agent,
            suggestedSkill: preprocessResult.routing.skill,
            confidence: preprocessResult.routing.confidence,
        };
    }
    async analyzeDelegation(request) {
        try {
            const metrics = {
                fileCount: request.context?.files?.length || 1,
                changeVolume: request.context?.changeVolume || 10,
                operationType: this.mapOperationToType(request.operation),
                dependencies: request.context?.dependencies?.length || 0,
                riskLevel: request.context?.riskLevel || "low",
                estimatedDuration: 10,
                operation: request.operation,
            };
            const complexityScore = this.complexityAnalyzer.calculateComplexityScore(metrics);
            // KERNEL ANALYSIS: Apply kernel pattern detection
            const kernelInsights = this.kernel.analyze(request.description);
            // Log kernel insights for debugging and learning
            frameworkLogger.log("agent-delegator", "kernel-analysis", "info", {
                operation: request.operation,
                kernelLevel: kernelInsights.level,
                kernelConfidence: kernelInsights.confidence,
                detectedPatterns: kernelInsights.cascadePatterns?.length || 0,
                detectedAssumptions: kernelInsights.fatalAssumptions?.length || 0,
            });
            const agentDetails = this.determineAgents(metrics, complexityScore, kernelInsights);
            const conflictResolution = this.determineConflictResolution(metrics, complexityScore);
            const agents = agentDetails.map((a) => a.name);
            // Persist delegation analysis metrics to state for tracking
            const existingMetrics = this.stateManager.get("delegation_metrics") || [];
            const delegationMetric = {
                timestamp: Date.now(),
                operation: request.operation,
                strategy: complexityScore.recommendedStrategy,
                complexity: complexityScore,
                estimatedDuration: metrics.estimatedDuration,
                agents: agents,
                analysisOnly: true,
                sessionId: request.sessionId,
            };
            existingMetrics.push(delegationMetric);
            this.stateManager.set("delegation_metrics", existingMetrics);
            await frameworkLogger.log("agent-delegator", "delegation-analyzed", "info", {
                operation: request.operation,
                strategy: complexityScore.recommendedStrategy,
                complexity: complexityScore.score,
                agentsCount: agents.length,
            }, request.sessionId);
            return {
                strategy: complexityScore.recommendedStrategy,
                agents,
                agentDetails,
                complexity: complexityScore,
                conflictResolution,
                estimatedDuration: metrics.estimatedDuration,
            };
        }
        catch (error) {
            await frameworkLogger.log("agent-delegator", "analysis-failed", "error", {
                operation: request.operation,
                error: String(error),
            }, request.sessionId);
            throw error;
        }
    }
    determineAgents(metrics, complexityScore, kernelInsights) {
        const agents = [];
        const operation = metrics.operation;
        if (operation === "security") {
            // KERNEL-AWARE: Apply P6 (Security Vulnerability) and A8/A9 patterns
            if (kernelInsights.fatalAssumptions?.some(a => a.id === 'A8' || a.id === 'A9')) {
                // Kernel detected security foundation assumption
                frameworkLogger.log("agent-delegator", "kernel-guided-security", "info", {
                    detectedPattern: kernelInsights.fatalAssumptions?.find(a => a.id)?.id,
                    guidance: 'Apply security foundation protocols first',
                    kernelAction: kernelInsights.fatalAssumptions?.find(a => a.id)?.action,
                });
            }
            agents.push({
                name: "security-auditor",
                confidence: 0.95,
                role: "security",
            });
        }
        if (operation === "review") {
            agents.push({ name: "code-reviewer", confidence: 0.8, role: "review" });
        }
        if (operation === "design") {
            agents.push({ name: "architect", confidence: 0.9, role: "design" });
        }
        if (complexityScore.recommendedStrategy === "multi-agent" ||
            complexityScore.recommendedStrategy === "orchestrator-led") {
            if (metrics.riskLevel === "critical") {
                agents.push({
                    name: "security-auditor",
                    confidence: 0.95,
                    role: "security",
                });
            }
            if (metrics.operationType === "refactor") {
                agents.push({
                    name: "refactorer",
                    confidence: 0.85,
                    role: "refactoring",
                });
            }
            if (metrics.dependencies > 3) {
                agents.push({ name: "code-reviewer", confidence: 0.8, role: "review" });
            }
            if (agents.length === 1) {
                agents.push({ name: "enforcer", confidence: 0.75, role: "validation" });
            }
        }
        else {
            if (metrics.operationType === "debug") {
                agents.push({
                    name: "bug-triage-specialist",
                    confidence: 0.85,
                    role: "debugging",
                });
            }
            else if (metrics.operationType === "test") {
                agents.push({
                    name: "testing-lead",
                    confidence: 0.8,
                    role: "testing",
                });
            }
            else if (metrics.operationType === "refactor") {
                agents.push({
                    name: "refactorer",
                    confidence: 0.85,
                    role: "refactoring",
                });
            }
            else if (metrics.operationType === "create" && operation !== "design") {
                agents.push({ name: "architect", confidence: 0.85, role: "design" });
            }
            else if (operation === "review" && agents.length === 0) {
                // Add code-reviewer for review tasks if not already added
                agents.push({ name: "code-reviewer", confidence: 0.8, role: "review" });
            }
            else if (complexityScore.level === "moderate" ||
                metrics.dependencies > 3) {
                agents.push({ name: "code-reviewer", confidence: 0.8, role: "review" });
            }
            else if (agents.length === 0) {
                agents.push({ name: "enforcer", confidence: 0.75, role: "validation" });
            }
        }
        if (complexityScore.recommendedStrategy === "orchestrator-led") {
            agents.push({
                name: "orchestrator",
                confidence: 0.9,
                role: "coordination",
            });
        }
        return agents.length > 0
            ? agents
            : [{ name: "enforcer", confidence: 0.75, role: "validation" }];
    }
    mapOperationToType(operation) {
        const op = operation.toLowerCase();
        if (op.includes("security") || op.includes("audit")) {
            return "analyze";
        }
        if (op.includes("review") || op.includes("quality")) {
            return "modify";
        }
        if (op.includes("refactor") || op.includes("restructure")) {
            return "refactor";
        }
        if (op.includes("debug") || op.includes("fix") || op.includes("resolve")) {
            return "debug";
        }
        if (op.includes("test")) {
            return "test";
        }
        if (op.includes("create") || op.includes("add") || op.includes("new")) {
            return "create";
        }
        if (op.includes("design") || op.includes("architect")) {
            return "create";
        }
        return "modify";
    }
    determineConflictResolution(metrics, complexityScore) {
        if (complexityScore.level === "enterprise" ||
            metrics.riskLevel === "critical") {
            return "expert_priority";
        }
        if (complexityScore.level === "complex" || metrics.dependencies > 5) {
            return "consensus";
        }
        return "majority_vote";
    }
    /**
   * Resolve the project directory for delegated agents
   * Ensures working directory is set correctly (src/ vs dist/ vs project root)
   */
    resolveProjectDirectory() {
        // Try to find the project root by looking for key files
        // Only use process.cwd() and validate it exists
        const root = process.cwd();
        const fs = require('fs');
        // Check if this looks like a valid project root
        if (fs.existsSync(`${root}/package.json`) || fs.existsSync(`${root}/strray.config.json`)) {
            return root;
        }
        // Fallback - try to find a parent with package.json
        let current = root;
        const maxLevels = 5;
        for (let i = 0; i < maxLevels; i++) {
            const parent = require('path').dirname(current);
            if (fs.existsSync(`${parent}/package.json`)) {
                return parent;
            }
            current = parent;
        }
        // Last resort fallback
        return root;
    }
    /**
     * Allowlist of valid agent names for dynamic imports
     * Prevents path traversal and unauthorized agent loading
     */
    static ALLOWED_AGENTS = new Set([
        'enforcer',
        'architect',
        'orchestrator',
        'bug-triage-specialist',
        'code-reviewer',
        'security-auditor',
        'refactorer',
        'testing-lead',
        'log-monitor',
        'researcher',
        'multimodal-looker',
        'analyzer',
        'seo-consultant',
        'content-creator',
        'growth-strategist',
        'database-engineer',
        'devops-engineer',
        'backend-engineer',
        'frontend-engineer',
        'frontend-ui-ux-engineer',
        'tech-writer',
        'performance-engineer',
        'mobile-developer',
        'strategist',
    ]);
    /**
     * Validate agent name for security
     * @throws Error if agent name is invalid or not allowed
     */
    validateAgentName(agentName) {
        // Check for path traversal attempts
        if (agentName.includes('..') || agentName.includes('/') || agentName.includes('\\')) {
            throw new Error(`Invalid agent name: path traversal detected in "${agentName}"`);
        }
        // Check for other dangerous characters
        if (!/^[a-zA-Z0-9_-]+$/.test(agentName)) {
            throw new Error(`Invalid agent name: only alphanumeric, underscore, and hyphen allowed in "${agentName}"`);
        }
        // Check allowlist
        if (!AgentDelegator.ALLOWED_AGENTS.has(agentName)) {
            throw new Error(`Agent "${agentName}" is not in the allowed list`);
        }
    }
    /**
     * Create a properly configured agent with full tool access and working directory context
     */
    createProperlyConfiguredAgent(agentName, agentConfig, request) {
        // Validate agent name before any operation (security)
        this.validateAgentName(agentName);
        // Create the agent with proper context
        return {
            config: agentConfig,
            agentName: agentName,
            execute: async (req) => {
                try {
                    // Import the specific agent implementation
                    const agentModule = await import(`../agents/${agentName}.js`);
                    const agentImplementation = agentModule[agentName];
                    if (agentImplementation && typeof agentImplementation.execute === 'function') {
                        // Ensure the agent has access to built-in OpenCode tools
                        const enhancedRequest = {
                            ...req,
                            context: {
                                ...req.context,
                                workingDirectory: process.cwd(),
                                availableTools: ['read', 'write', 'edit', 'glob', 'grep', 'bash', 'task', 'webfetch', 'todowrite', 'todoread', 'skill'],
                                isDelegated: true
                            }
                        };
                        return await agentImplementation.execute(enhancedRequest);
                    }
                    else {
                        throw new Error(`Agent ${agentName} does not have a valid execute method`);
                    }
                }
                catch (error) {
                    // Fallback to structured error response for agents without proper implementations
                    return {
                        agent: agentName,
                        operation: request.operation,
                        description: request.description,
                        capabilities: agentConfig.capabilities,
                        mode: agentConfig.mode,
                        status: "error",
                        timestamp: new Date().toISOString(),
                        error: `Agent execution failed: ${String(error)}`,
                        recommendations: [
                            "Check if agent implementation exists",
                            "Verify agent has proper execute method",
                            "Ensure agent supports delegated mode"
                        ]
                    };
                }
            },
            getCapabilities: () => agentConfig.capabilities,
            getMaxComplexity: () => agentConfig.maxComplexity,
            isEnabled: () => agentConfig.enabled,
            isAgent: true, // Distinction from skills
            getType: () => 'agent'
        };
    }
    /**
     * Clear agent-stub creation and deprecated features
     * Remove this method if it exists in the future version
     */
    clearDeprecatedStubAgent() {
        // Legacy method removal - this ensures no stub agents are created
        frameworkLogger.log("agent-delegator", "deprecated-method-removed", "info", {
            message: "Cleared deprecated stub agent creation to ensure proper tool access",
            timestamp: new Date().toISOString()
        });
    }
    async executeDelegation(analysis, request) {
        const startTime = Date.now();
        const results = [];
        const errors = [];
        // Set up proper working directory for delegated agents
        const originalWorkingDir = process.cwd();
        const projectRootDir = this.resolveProjectDirectory();
        try {
            // Change to project working directory for all delegated agents
            process.chdir(projectRootDir);
            // Properly configured agent execution
            const { builtinAgents } = await import("../agents/index.js");
            for (const agentName of analysis.agents) {
                const agentStartTime = Date.now();
                try {
                    // First check state manager for custom/mock agents (for testing)
                    let agentInstance = this.stateManager.get(`agent:${agentName}`);
                    // If no custom agent in state, use builtin configuration
                    if (!agentInstance) {
                        const agentConfig = builtinAgents[agentName];
                        if (!agentConfig) {
                            throw new Error(`Agent ${agentName} not found in builtin agents`);
                        }
                        // Create properly configured agent with full tool access
                        agentInstance = this.createProperlyConfiguredAgent(agentName, agentConfig, request);
                    }
                    // Execute the agent
                    const output = await agentInstance.execute(request);
                    const executionTime = Date.now() - agentStartTime;
                    results.push({
                        agent: agentName,
                        output,
                        executionTime,
                    });
                    await frameworkLogger.log("agent-delegator", "agent-executed", "success", {
                        agent: agentName,
                        executionTime,
                        success: true,
                        workingDirectory: projectRootDir,
                    }, request.sessionId);
                }
                catch (error) {
                    const executionTime = Date.now() - agentStartTime;
                    const errorMessage = `Agent ${agentName} failed: ${String(error)}`;
                    errors.push(errorMessage);
                    await frameworkLogger.log("agent-delegator", "agent-execution-failed", "error", {
                        agent: agentName,
                        executionTime,
                        error: errorMessage,
                        workingDirectory: projectRootDir,
                    }, request.sessionId);
                }
            }
            const totalTime = Date.now() - startTime;
            const success = errors.length === 0 && results.length > 0;
            // Persist delegation execution metrics to state for tracking
            const existingMetrics = this.stateManager.get("delegation_metrics") || [];
            const delegationMetric = {
                timestamp: Date.now(),
                operation: request.operation,
                strategy: analysis.strategy,
                complexity: analysis.complexity,
                estimatedDuration: analysis.estimatedDuration,
                agents: analysis.agents,
                results,
                totalTime,
                success,
                errors,
                analysisOnly: false,
                sessionId: request.sessionId,
            };
            existingMetrics.push(delegationMetric);
            this.stateManager.set("delegation_metrics", existingMetrics);
            await frameworkLogger.log("agent-delegator", "delegation-completed", success ? "success" : "error", {
                strategy: analysis.strategy,
                agentsCount: analysis.agents.length,
                successCount: results.length,
                errorCount: errors.length,
                totalTime,
            }, request.sessionId);
            // Restore original working directory
            process.chdir(originalWorkingDir);
            return {
                success,
                results,
                totalTime,
                errors: errors.length > 0 ? errors : undefined,
            };
        }
        catch (error) {
            // Ensure working directory is restored even on failure
            process.chdir(originalWorkingDir);
            const totalTime = Date.now() - startTime;
            await frameworkLogger.log("agent-delegator", "delegation-failed", "error", {
                error: String(error),
                totalTime,
                workingDirectoryRestored: process.cwd(),
            }, request.sessionId);
            throw error;
        }
    }
    getPerformanceMetrics() {
        const delegations = this.stateManager.get("delegation_metrics") || [];
        const executionDelegations = delegations.filter((d) => !d.analysisOnly);
        const totalDelegations = executionDelegations.length > 0
            ? executionDelegations.length
            : delegations.length;
        const successfulDelegations = executionDelegations.filter((d) => d.success).length;
        const failedDelegations = executionDelegations.filter((d) => !d.success).length;
        const averageExecutionTime = totalDelegations > 0
            ? delegations.reduce((sum, d) => sum + d.totalTime, 0) /
                totalDelegations
            : 0;
        const agentUtilization = {};
        const strategyUsage = {};
        let totalComplexity = 0;
        let totalDuration = 0;
        delegations.forEach((delegation) => {
            // Count strategy usage for both analyses and executions
            if (delegation.strategy) {
                strategyUsage[delegation.strategy] =
                    (strategyUsage[delegation.strategy] || 0) + 1;
            }
            if (delegation.complexity?.score) {
                totalComplexity += delegation.complexity.score;
            }
            if (delegation.estimatedDuration) {
                totalDuration += delegation.estimatedDuration;
            }
            delegation.results?.forEach((result) => {
                agentUtilization[result.agent] =
                    (agentUtilization[result.agent] || 0) + 1;
            });
        });
        const averageResponseTime = averageExecutionTime;
        const averageComplexity = totalDelegations > 0 ? totalComplexity / totalDelegations : 0;
        const averageDuration = totalDelegations > 0 ? totalDuration / totalDelegations : 0;
        return {
            totalDelegations,
            successfulDelegations,
            failedDelegations,
            averageExecutionTime,
            averageResponseTime,
            averageComplexity,
            averageDuration,
            strategyUsage,
            agentUtilization,
        };
    }
    getDelegationMetrics() {
        const baseMetrics = this.getPerformanceMetrics();
        const recentDelegations = (this.stateManager.get("delegation_metrics") || [])
            .slice(-10)
            .map((d) => ({
            timestamp: d.timestamp || Date.now(),
            operation: d.operation,
            strategy: d.strategy,
            success: d.success,
            totalTime: d.totalTime,
        }));
        return {
            ...baseMetrics,
            recentDelegations,
        };
    }
    updateAgentCapability(agentName, capabilities) {
        const availableAgents = this.getAvailableAgents();
        const agentIndex = availableAgents.findIndex((a) => a.name === agentName);
        if (agentIndex !== -1) {
            const existingAgent = availableAgents[agentIndex];
            const updatedAgent = {
                ...existingAgent,
                ...capabilities,
                name: agentName,
                capabilities: capabilities.capabilities || existingAgent?.capabilities || [],
                status: capabilities.status || existingAgent?.status || "active",
            };
            availableAgents[agentIndex] = updatedAgent;
            this.stateManager.set(`agent_capabilities:${agentName}`, updatedAgent);
            frameworkLogger.log("agent-delegator", "agent-capability-updated", "info", {
                agentName,
                updatedCapabilities: capabilities,
            });
        }
    }
}
export function createAgentDelegator(stateManager, configLoader) {
    return new AgentDelegator(stateManager, configLoader);
}
