// strray-ai/dist/processors/implementations/coverage-analysis-processor.js
/**
 * Coverage Analysis Processor
 *
 * Post-processor that analyzes test coverage.
 *
 * @module processors/implementations
 * @version 1.0.0
 */
import { PostProcessor } from "../processor-interfaces.js";
import { frameworkLogger } from "../../core/framework-logger.js";
export class CoverageAnalysisProcessor extends PostProcessor {
    name = "coverageAnalysis";
    priority = 65;
    async run(context) {
        const ctx = context;
        await frameworkLogger.log("coverage-analysis-processor", "analyzing", "info", {
            operation: ctx.operation,
            filePath: this.getFilePath(ctx)?.slice(0, 100),
        });
        // Coverage analysis is informational - return success even if no coverage data
        // This processor serves as a hook point for future coverage analysis implementations
        return {
            success: true,
            message: "Coverage analysis skipped - no coverage data available",
            coverage: null,
            analyzedAt: new Date().toISOString(),
        };
    }
}
