// strray-ai/dist/delegation/routing/history-matcher.js
/**
 * History Matcher
 *
 * Extracted history-based routing logic from task-skill-router.ts
 * Phase 3 refactoring - Matching Logic Extraction
 */
/**
 * History matcher for routing based on past routing outcomes
 */
export class HistoryMatcher {
    history = new Map();
    minHistorySuccessRate;
    minAttempts;
    /**
     * Create a new HistoryMatcher
     * @param minSuccessRate - Minimum success rate to consider (default: 0.7)
     * @param minAttempts - Minimum number of attempts required (default: 3)
     */
    constructor(minSuccessRate = 0.7, minAttempts = 3) {
        this.minHistorySuccessRate = minSuccessRate;
        this.minAttempts = minAttempts;
    }
    /**
     * Match task by historical routing data
     * @param taskId - Task ID to look up in history
     * @returns Routing result or null if no suitable history found
     */
    match(taskId) {
        const entry = this.history.get(taskId);
        if (entry && entry.totalAttempts >= this.minAttempts) {
            const successRate = entry.successCount / entry.totalAttempts;
            if (successRate >= this.minHistorySuccessRate) {
                return {
                    skill: entry.skill,
                    agent: entry.agent,
                    confidence: successRate,
                    fromHistory: true,
                    reason: `History: ${(successRate * 100).toFixed(0)}% success rate (${entry.successCount}/${entry.totalAttempts})`,
                };
            }
        }
        return null;
    }
    /**
     * Track routing result for learning
     * @param taskId - Task ID
     * @param agent - Agent that was routed to
     * @param skill - Skill that was used
     * @param success - Whether the routing was successful
     */
    track(taskId, agent, skill, success) {
        const entry = this.history.get(taskId);
        if (entry) {
            // Update existing entry
            entry.totalAttempts++;
            if (success) {
                entry.successCount++;
            }
            entry.lastUsed = new Date();
        }
        else {
            // Create new entry
            this.history.set(taskId, {
                agent,
                skill,
                successCount: success ? 1 : 0,
                totalAttempts: 1,
                lastUsed: new Date(),
            });
        }
    }
    /**
     * Track routing result (alias for track method, for backward compatibility)
     * @param taskId - Task ID
     * @param agent - Agent that was routed to
     * @param skill - Skill that was used
     * @param success - Whether the routing was successful
     */
    trackResult(taskId, agent, skill, success) {
        this.track(taskId, agent, skill, success);
    }
    /**
     * Get statistics for all tracked tasks
     * @returns Array of history statistics
     */
    getStats() {
        return Array.from(this.history.entries()).map(([taskId, entry]) => ({
            taskId,
            successRate: entry.totalAttempts > 0 ? entry.successCount / entry.totalAttempts : 0,
            count: entry.totalAttempts,
            agent: entry.agent,
        }));
    }
    /**
     * Get statistics for a specific task
     * @param taskId - Task ID to get stats for
     * @returns History stats or null if not found
     */
    getTaskStats(taskId) {
        const entry = this.history.get(taskId);
        if (!entry) {
            return null;
        }
        return {
            taskId,
            successRate: entry.totalAttempts > 0 ? entry.successCount / entry.totalAttempts : 0,
            count: entry.totalAttempts,
            agent: entry.agent,
        };
    }
    /**
     * Clear all history
     */
    clear() {
        this.history.clear();
    }
    /**
     * Clear history for a specific task
     * @param taskId - Task ID to clear
     */
    clearTask(taskId) {
        this.history.delete(taskId);
    }
    /**
     * Get number of tracked tasks
     * @returns Count of history entries
     */
    size() {
        return this.history.size;
    }
    /**
     * Load history from a plain object (for state restoration)
     * @param data - History data to load
     */
    loadHistory(data) {
        this.history.clear();
        for (const [taskId, entry] of Object.entries(data)) {
            // Convert date string back to Date object
            this.history.set(taskId, {
                ...entry,
                lastUsed: new Date(entry.lastUsed),
            });
        }
    }
    /**
     * Export history to a plain object (for state persistence)
     * @returns History data as plain object
     */
    exportHistory() {
        const data = {};
        for (const [taskId, entry] of this.history) {
            data[taskId] = entry;
        }
        return data;
    }
    /**
     * Set minimum success rate threshold
     * @param rate - New minimum success rate
     */
    setMinSuccessRate(rate) {
        this.minHistorySuccessRate = rate;
    }
    /**
     * Set minimum attempts threshold
     * @param attempts - New minimum attempts
     */
    setMinAttempts(attempts) {
        this.minAttempts = attempts;
    }
    /**
     * Get top performing agents based on history
     * @param limit - Maximum number of results (default: 5)
     * @returns Array of top agents with success rates
     */
    getTopAgents(limit = 5) {
        const agentStats = new Map();
        for (const entry of this.history.values()) {
            const stats = agentStats.get(entry.agent) || { successes: 0, total: 0 };
            stats.successes += entry.successCount;
            stats.total += entry.totalAttempts;
            agentStats.set(entry.agent, stats);
        }
        return Array.from(agentStats.entries())
            .map(([agent, stats]) => ({
            agent,
            successRate: stats.total > 0 ? stats.successes / stats.total : 0,
            count: stats.total,
        }))
            .filter(a => a.count >= this.minAttempts)
            .sort((a, b) => b.successRate - a.successRate)
            .slice(0, limit);
    }
}
