// strray-ai/dist/mcps/tools/tool-registry.js
/**
 * Tool Registry
 *
 * Manages tool registration and lookup across MCP servers.
 * Part of Phase 4 refactoring - Tool Layer extraction.
 */
export class ToolRegistry {
    tools = new Map();
    /**
     * Register tools for a specific server
     */
    register(serverName, tools) {
        if (!this.tools.has(serverName)) {
            this.tools.set(serverName, new Map());
        }
        const serverTools = this.tools.get(serverName);
        for (const tool of tools) {
            serverTools.set(tool.name, tool);
        }
    }
    /**
     * Get a specific tool by server and tool name
     */
    getTool(serverName, toolName) {
        return this.tools.get(serverName)?.get(toolName);
    }
    /**
     * Get all tools for a specific server
     */
    getTools(serverName) {
        return Array.from(this.tools.get(serverName)?.values() || []);
    }
    /**
     * Check if a tool exists
     */
    hasTool(serverName, toolName) {
        return this.tools.get(serverName)?.has(toolName) || false;
    }
    /**
     * Clear all registered tools
     */
    clear() {
        this.tools.clear();
    }
    /**
     * Get all registered server names
     */
    getServerNames() {
        return Array.from(this.tools.keys());
    }
    /**
     * Remove all tools for a specific server
     */
    unregisterServer(serverName) {
        return this.tools.delete(serverName);
    }
    /**
     * Get total tool count across all servers
     */
    getToolCount() {
        let count = 0;
        for (const serverTools of this.tools.values()) {
            count += serverTools.size;
        }
        return count;
    }
}
