// strray-ai/dist/reporting/framework-reporting-system.js
import { frameworkLogger } from "../core/framework-logger.js";
import * as fs from "fs";
import * as path from "path";
import { getComprehensiveLogs, filterLogsByConfig, frameworkLogToParsedEntry, } from "./log-parser.js";
import { calculateTimeRange, calculateMetrics, calculatePeakActivity, calculateHealthScore, generateInsights, generateRecommendations, generateAlerts, } from "./metrics.js";
import { formatReport } from "./report-formatter.js";
export class FrameworkReportingSystem {
    logRetentionHours = 24;
    reportCache = new Map();
    async generateReport(config) {
        const reportId = this.generateReportId(config);
        const cached = this.getCachedReport(reportId);
        if (cached && this.isCacheValid(cached.timestamp)) {
            return formatReport(cached.data, config.outputFormat);
        }
        const reportData = await this.collectReportData(config);
        this.cacheReport(reportId, reportData);
        const formattedReport = formatReport(reportData, config.outputFormat);
        if (config.outputPath) {
            await this.saveReportToFile(formattedReport, config.outputPath);
        }
        return formattedReport;
    }
    scheduleAutomatedReports(schedule) {
        const intervalMs = this.getIntervalMs(schedule.frequency);
        setInterval(async () => {
            for (const type of schedule.types) {
                const config = {
                    type,
                    timeRange: { lastHours: schedule.frequency === "hourly" ? 1 : 24 },
                    outputFormat: "markdown",
                    outputPath: path.join(schedule.outputDir, `${type}-report-${new Date().toISOString().split("T")[0]}.md`),
                };
                try {
                    await this.generateReport(config);
                    await frameworkLogger.log("framework-reporting-system", "automated-report-generated", "success", { reportType: type, outputPath: config.outputPath });
                }
                catch (error) {
                    await frameworkLogger.log("framework-reporting-system", "automated-report-generation-failed", "error", { reportType: type, error: String(error) });
                }
            }
            await this.cleanupOldReports(schedule.outputDir, schedule.retentionDays);
        }, intervalMs);
    }
    async getRealtimeStatus() {
        const recentLogs = frameworkLogger.getRecentLogs(50);
        const parsedLogs = recentLogs.map(frameworkLogToParsedEntry);
        const activeComponents = [
            ...new Set(recentLogs.map((log) => log.component)),
        ];
        const recentActivity = parsedLogs.slice(0, 10);
        const errorCount = recentLogs.filter((log) => log.status === "error").length;
        const successCount = recentLogs.filter((log) => log.status === "success").length;
        const healthScore = recentLogs.length > 0
            ? (successCount / (successCount + errorCount)) * 100
            : 100;
        const alerts = generateAlerts(parsedLogs);
        return {
            activeComponents,
            recentActivity,
            healthScore,
            alerts,
        };
    }
    createCustomReport(template) {
        return `
# Custom Report: ${template.name}

## Filters Applied
${JSON.stringify(template.filters, null, 2)}

## Aggregations
${JSON.stringify(template.aggregations, null, 2)}

## Visualizations
${template.visualizations.join(", ")}

## Usage
\`\`\`typescript
const report = await reportingSystem.generateCustomReport('${template.name}');
\`\`\`
    `;
    }
    generateReportId(config) {
        const timeKey = config.timeRange?.lastHours || "all";
        return `${config.type}-${config.outputFormat}-${timeKey}`;
    }
    getCachedReport(reportId) {
        return this.reportCache.get(reportId) || null;
    }
    isCacheValid(timestamp) {
        const cacheAgeMs = Date.now() - timestamp.getTime();
        const maxCacheAgeMs = 5 * 60 * 1000;
        return cacheAgeMs < maxCacheAgeMs;
    }
    cacheReport(reportId, data) {
        this.reportCache.set(reportId, { data, timestamp: new Date() });
        if (this.reportCache.size > 10) {
            const keys = Array.from(this.reportCache.keys());
            if (keys.length > 0) {
                this.reportCache.delete(keys[0]);
            }
        }
    }
    async collectReportData(config) {
        const logs = filterLogsByConfig(await getComprehensiveLogs(config), config);
        const timeRange = calculateTimeRange(logs, config.timeRange);
        const metrics = calculateMetrics(logs);
        const chronologicalEvents = logs;
        const insights = generateInsights(logs, metrics);
        const recommendations = generateRecommendations(metrics);
        const summary = {
            totalEvents: logs.length,
            activeComponents: [...new Set(logs.map((log) => log.component))],
            peakActivity: calculatePeakActivity(logs),
            healthScore: calculateHealthScore(logs),
        };
        return {
            generatedAt: new Date(),
            timeRange,
            metrics,
            chronologicalEvents,
            insights,
            recommendations,
            summary,
        };
    }
    async saveReportToFile(content, filePath) {
        const dir = path.dirname(filePath);
        await fs.promises.mkdir(dir, { recursive: true });
        await fs.promises.writeFile(filePath, content, "utf8");
    }
    async cleanupOldReports(outputDir, retentionDays) {
        try {
            const files = await fs.promises.readdir(outputDir);
            const cutoffTime = Date.now() - retentionDays * 24 * 60 * 60 * 1000;
            for (const file of files) {
                const filePath = path.join(outputDir, file);
                const stats = await fs.promises.stat(filePath);
                if (stats.mtime.getTime() < cutoffTime) {
                    await fs.promises.unlink(filePath);
                }
            }
        }
        catch {
            // Ignore cleanup errors
        }
    }
    getIntervalMs(frequency) {
        switch (frequency) {
            case "hourly":
                return 60 * 60 * 1000;
            case "daily":
                return 24 * 60 * 60 * 1000;
            case "weekly":
                return 7 * 24 * 60 * 60 * 1000;
        }
    }
}
export const frameworkReportingSystem = new FrameworkReportingSystem();
if (import.meta.url === `file://${process.argv[1]}`) {
    const args = process.argv.slice(2);
    const reportType = args[0] || "full-analysis";
    const outputFormat = args[1] || "markdown";
    const config = {
        type: reportType,
        outputFormat,
        outputPath: args[2] ||
            `./reports/${reportType}-report-${new Date().toISOString().split("T")[0]}.${outputFormat === "markdown" ? "md" : outputFormat}`,
        detailedMetrics: true,
    };
    frameworkReportingSystem
        .generateReport(config)
        .then(async (result) => {
        await frameworkLogger.log("framework-reporting-system", "report-generated-successfully", "success", { message: "Report generated successfully" });
        if (config.outputPath) {
            await frameworkLogger.log("framework-reporting-system", "saved-to-output-path", "info", { message: `Saved to: ${config.outputPath}` });
        }
        else {
            await frameworkLogger.log("framework-reporting-system", "result", "info", { message: result });
        }
    })
        .catch(async (error) => {
        await frameworkLogger.log("framework-reporting-system", "report-generation-failed", "error", { error: String(error) });
        process.exit(1);
    });
}
