// strray-ai/dist/delegation/config/default-mappings/devops-mappings.js
/**
 * DevOps Mappings
 *
 * Maps DevOps and deployment tasks to devops-engineer and architect agents.
 * Includes CI/CD, Docker, Kubernetes, and infrastructure tasks.
 */
export const DEVOPS_MAPPINGS = [
    {
        keywords: ["devops", "docker", "kubernetes", "ci/cd", "pipeline", "deploy"],
        skill: "devops-automation",
        agent: "devops-engineer",
        confidence: 0.95,
        category: "devops",
        priority: "high",
    },
    {
        keywords: ["monitor", "log", "alert", "metrics", "observability"],
        skill: "log-monitoring",
        agent: "log-monitor",
        confidence: 0.95,
        category: "devops",
        priority: "high",
    },
    {
        keywords: [
            // === ACTION VERB + DEVOPS PATTERNS ===
            "deploy code", "deploy application", "deploy to production", "deploy to staging",
            "setup ci/cd", "build pipeline", "create pipeline", "configure pipeline",
            "setup kubernetes", "deploy kubernetes", "configure k8s", "setup docker",
            "build docker", "dockerize application", "create dockerfile", "setup deployment",
            "release", "publish release", "tag release", "create release",
            // === DEVOPS OBJECTS ===
            "deploy", "kubernetes", "k8s", "docker", "dockerfile", "ci/cd", "pipeline",
            "release", "version", "semver", "git tag",
            "production", "staging", "environment", "infrastructure", "deployment",
        ],
        skill: "devops-deployment",
        agent: "architect",
        confidence: 0.85,
        category: "devops",
        priority: "high",
    },
    {
        keywords: ["log", "logging", "monitor", "alert", "observability"],
        skill: "log-monitor",
        agent: "log-monitor",
        confidence: 0.9,
        category: "devops",
        priority: "medium",
    },
    {
        keywords: [
            "aws",
            "lambda",
            "serverless",
            "s3",
            "dynamodb",
            "cloudformation",
        ],
        skill: "aws-serverless",
        agent: "devops-engineer",
        confidence: 0.85,
        category: "devops",
        priority: "medium",
    },
    {
        keywords: ["boot", "init", "startup", "initialize", "setup", "config"],
        skill: "boot-orchestrator",
        agent: "architect",
        confidence: 0.9,
        category: "devops",
        priority: "medium",
    },
    {
        keywords: ["pipeline", "batch", "stream", "transform", "filter", "etl"],
        skill: "processor-pipeline",
        agent: "architect",
        confidence: 0.85,
        category: "devops",
        priority: "medium",
    },
];
