// strray-ai/dist/delegation/config/default-mappings/performance-mappings.js
/**
 * Performance Mappings
 *
 * Maps performance optimization tasks to performance-related agents.
 * Includes optimization, bottleneck detection, and memory/CPU tuning.
 */
export const PERFORMANCE_MAPPINGS = [
    {
        keywords: [
            // === ACTION VERB + PERFORMANCE PATTERNS ===
            "optimize performance", "improve performance", "speed up code", "reduce latency",
            "fix memory leak", "optimize cpu", "improve throughput",
            // === Object-focused patterns
            "bottleneck", "memory leak", "cpu usage", "latency", "throughput",
        ],
        skill: "performance-optimization",
        agent: "mobile-developer",
        confidence: 0.98,
        category: "performance",
        priority: "high",
    },
    {
        keywords: [
            "optimize", "speed up", "improve", "enhance", "accelerate",
            "performance", "slow", "fast", "efficient",
        ],
        skill: "performance-optimization",
        agent: "performance-engineer",
        confidence: 0.9,
        category: "performance",
        priority: "high",
    },
    {
        keywords: [
            "optimize", "speed up", "improve", "enhance", "accelerate",
            "slow", "fast", "efficient",
        ],
        skill: "performance-optimization",
        agent: "performance-engineer",
        confidence: 0.8,
        category: "performance",
        priority: "medium",
    },
    {
        keywords: [
            // === ACTION VERB + PERFORMANCE PATTERNS ===
            "optimize performance", "improve performance", "speed up code", "reduce latency",
            "fix memory leak", "optimize cpu", "improve throughput",
            // === Object-focused patterns
            "bottleneck", "memory leak", "cpu usage", "latency", "throughput",
        ],
        skill: "performance-optimization",
        agent: "performance-engineer",
        confidence: 0.9,
        category: "performance",
        priority: "high",
    },
    // SPECIAL CASE handled in special-mappings.ts
    // {
    //   keywords: ["improve application performance", "speed up application"],
    //   skill: "performance-optimization",
    //   agent: "mobile-developer",
    //   confidence: 0.99,
    //   category: "performance",
    //   priority: "critical",
    // },
];
