// strray-ai/dist/delegation/config/default-mappings/security-mappings.js
/**
 * Security Mappings
 *
 * Maps security-related tasks to the security-auditor agent.
 * Includes vulnerability scanning, authentication, authorization, and OWASP concerns.
 */
export const SECURITY_MAPPINGS = [
    {
        keywords: [
            // === ACTION VERB + SECURITY PATTERNS ===
            "audit security", "check security", "scan security", "review security", "analyze security",
            "fix vulnerability", "patch vulnerability", "resolve security issue",
            "encrypt data", "secure data", "sanitize input", "validate credentials",
            "implement auth", "add authentication", "setup authorization", "configure jwt",
            // === SECURITY CONCEPTS ===
            "security", "vulnerability", "vulnerabilities",
            "audit", "credential", "encrypt", "sanitize", "secure",
            "authentication", "authorization", "oauth", "jwt", "api key",
            "cve", "exploit", "penetration", "security scan", "pen test",
            // === OWASP CONCEPTS ===
            "injection", "xss", "csrf", "sql injection", "xxe", "ssrf",
            "broken authentication", "sensitive data exposure", "security misconfiguration",
        ],
        skill: "security-audit",
        agent: "security-auditor",
        confidence: 0.95,
        category: "security",
        priority: "high",
    },
    {
        keywords: [
            "vulnerability",
            "cve",
            "exploit",
            "penetration",
            "security scan",
        ],
        skill: "vulnerability-scanner",
        agent: "security-auditor",
        confidence: 0.9,
        category: "security",
        priority: "high",
    },
    {
        keywords: [
            "api security",
            "authentication",
            "authorization",
            "jwt",
            "oauth",
        ],
        skill: "api-security-best-practices",
        agent: "security-auditor",
        confidence: 0.9,
        category: "security",
        priority: "high",
    },
];
