// strray-ai/dist/delegation/config/default-mappings/ui-ux-mappings.js
/**
 * UI/UX Design Mappings
 *
 * Maps UI/UX design tasks to the frontend-ui-ux-engineer agent.
 * Includes component design, styling, layout, and interface creation.
 */
export const UI_UX_MAPPINGS = [
    {
        keywords: [
            // === ACTION VERB + OBJECT PATTERNS (most specific) ===
            "design component", "create component", "build component", "make component", "write component", "implement component",
            "design button", "create button", "build button", "make button", "add button", "implement button",
            "design form", "create form", "build form", "make form", "write form", "implement form",
            "design modal", "create modal", "build modal", "make modal", "add modal", "implement modal",
            "design page", "create page", "build page", "make page", "write page", "implement page",
            "design layout", "create layout", "build layout", "make layout", "implement layout",
            "design interface", "create interface", "build interface", "make interface", "implement interface",
            // === UI OBJECTS (ACTION-ORIENTED ONLY) ===
            "component", "button", "form", "modal", "dialog", "dropdown", "input", "textarea",
            "page", "layout", "interface", "header", "footer", "sidebar", "navbar",
            "card", "list", "table", "grid", "flex", "container", "wrapper",
            "design system", "component library", "ui kit", "style guide",
            // === UI DESIGN KEYWORDS ===
            "ui design", "ux design", "user interface", "user experience",
            "figma", "sketch", "mockup", "wireframe", "prototype", "design",
            "css", "styling", "responsive design", "mobile ui", "desktop ui",
            // === FRONTEND TECH ===
            "react component", "vue component", "angular component",
            "css component", "html element", "styled component",
        ],
        skill: "ui-ux-design",
        agent: "frontend-ui-ux-engineer",
        confidence: 0.98,
        category: "ui-ux",
        priority: "high",
    },
    {
        keywords: [
            // === ACTION VERB + PATTERNS ===
            "design ui", "design frontend", "build ui", "create ui", "implement ui",
            "design css", "build css", "create css", "implement css",
            "design page", "create page", "build page", "make page",
            "design interface", "create interface", "build interface", "make interface",
            // === UI OBJECTS ===
            "component", "button", "form", "modal", "dialog", "dropdown", "input", "textarea",
            "page", "layout", "interface", "header", "footer", "sidebar", "navbar",
        ],
        skill: "ui-ux-design",
        agent: "frontend-ui-ux-engineer",
        confidence: 0.75,
        category: "ui-ux",
        priority: "medium",
    },
    {
        keywords: [
            "ui",
            "frontend",
            "css",
            "button",
            "form",
            "modal",
            "page",
            "interface",
        ],
        skill: "ui-ux-design",
        agent: "enforcer",
        confidence: 0.75,
        category: "ui-ux",
        priority: "low",
    },
];
