// strray-ai/dist/enforcement/loaders/base-loader.js
/**
 * Base Rule Loader
 *
 * Abstract base class for all rule loaders. Provides common functionality
 * for loading rules from various sources such as files, APIs, or databases.
 *
 * Phase 4 refactoring: Extracted from RuleEnforcer to separate loading logic
 * from rule execution.
 *
 * @module loaders/base-loader
 * @version 1.0.0
 */
import * as fs from "fs";
import * as path from "path";
/**
 * Abstract base class implementing IRuleLoader interface.
 * Provides common file loading utilities and error handling.
 *
 * @example
 * ```typescript
 * class MyLoader extends BaseLoader {
 *   readonly name = 'my-loader';
 *
 *   async load(): Promise<RuleDefinition[]> {
 *     const data = await this.loadJsonFile('rules.json');
 *     return this.transformToRules(data);
 *   }
 *
 *   async isAvailable(): Promise<boolean> {
 *     return this.fileExists('rules.json');
 *   }
 * }
 * ```
 */
export class BaseLoader {
    /**
     * Load and parse a JSON file.
     * @param filePath - Path to the JSON file
     * @returns Promise resolving to parsed JSON data
     * @throws Error if file cannot be read or parsed
     */
    async loadJsonFile(filePath) {
        const content = await fs.promises.readFile(filePath, "utf8");
        return JSON.parse(content);
    }
    /**
     * Check if a file exists.
     * @param filePath - Path to check
     * @returns Promise resolving to true if file exists
     */
    async fileExists(filePath) {
        try {
            await fs.promises.access(filePath, fs.constants.F_OK);
            return true;
        }
        catch {
            return false;
        }
    }
    /**
     * Read file content as string.
     * @param filePath - Path to the file
     * @returns Promise resolving to file content
     * @throws Error if file cannot be read
     */
    async readFile(filePath) {
        return fs.promises.readFile(filePath, "utf8");
    }
    /**
     * Resolve a path relative to the current working directory.
     * @param relativePath - Path relative to process.cwd()
     * @returns Absolute path
     */
    resolvePath(relativePath) {
        return path.join(process.cwd(), relativePath);
    }
    /**
     * Get file stats.
     * @param filePath - Path to the file
     * @returns Promise resolving to file stats
     * @throws Error if file cannot be accessed
     */
    async getFileStats(filePath) {
        return fs.promises.stat(filePath);
    }
}
