// strray-ai/dist/enforcement/loaders/processor-loader.js
/**
 * Processor Rule Loader
 *
 * Loads processor-specific rules for the enforcement system.
 * This is a placeholder for future expansion of processor-specific validation.
 *
 * Phase 4 refactoring: Extracted from RuleEnforcer.loadProcessorRules()
 *
 * @module loaders/processor-loader
 * @version 1.0.0
 */
import { frameworkLogger } from "../../core/framework-logger.js";
import { BaseLoader } from "./base-loader.js";
/**
 * Loader for processor-specific rules.
 * Currently provides a placeholder for future processor rule expansion.
 *
 * @example
 * ```typescript
 * const loader = new ProcessorLoader();
 * const rules = await loader.load();
 * console.log(`Loaded ${rules.length} processor rules`);
 * ```
 */
export class ProcessorLoader extends BaseLoader {
    name = "processor";
    /**
     * Processor rules are always available.
     * @returns Always returns true
     */
    async isAvailable() {
        return true;
    }
    /**
     * Load processor-specific rules.
     * Currently returns placeholder rules for future expansion.
     * @returns Promise resolving to array of rule definitions
     */
    async load() {
        const rules = [];
        try {
            // Add processor validation rule
            rules.push(this.createProcessorValidationRule());
            // Add processor health check rule
            rules.push(this.createProcessorHealthRule());
            await frameworkLogger.log("processor-loader", "loaded-processor-rules", "success", {
                message: `Loaded ${rules.length} processor rules`,
                ruleCount: rules.length,
            });
        }
        catch (error) {
            await frameworkLogger.log("processor-loader", "failed-to-load-processor", "error", {
                message: `Failed to load processor rules: ${error instanceof Error ? error.message : String(error)}`,
                error: error instanceof Error ? error.message : String(error),
            });
        }
        return rules;
    }
    /**
     * Create processor validation rule.
     * @returns RuleDefinition for processor validation
     */
    createProcessorValidationRule() {
        return {
            id: "processor-validation",
            name: "Processor Validation",
            description: "Validates that processor operations follow established patterns",
            category: "architecture",
            severity: "info",
            enabled: true,
            validator: this.validateProcessorOperations.bind(this),
        };
    }
    /**
     * Create processor health check rule.
     * @returns RuleDefinition for processor health
     */
    createProcessorHealthRule() {
        return {
            id: "processor-health",
            name: "Processor Health Check",
            description: "Monitors processor health and performance metrics",
            category: "performance",
            severity: "info",
            enabled: true,
            validator: this.validateProcessorHealth.bind(this),
        };
    }
    /**
     * Validate processor operations.
     * @param context - Validation context
     * @returns Validation result
     */
    async validateProcessorOperations(context) {
        // Placeholder for processor operation validation
        return {
            passed: true,
            message: "Processor operations validated",
        };
    }
    /**
     * Validate processor health.
     * @param context - Validation context
     * @returns Validation result
     */
    async validateProcessorHealth(context) {
        // Placeholder for processor health validation
        return {
            passed: true,
            message: "Processor health check passed",
        };
    }
}
