// strray-ai/dist/security/secure-authentication-system.js
/**
 * Secure Authentication System
 *
 * Enterprise-grade authentication and authorization with JWT, OAuth2, and RBAC support.
 * Implements zero-trust security principles with comprehensive audit logging.
 *
 * @version 1.0.0
 * @since 2026-01-08
 */
import { EventEmitter } from "events";
import * as crypto from "crypto";
import * as jwt from "jsonwebtoken";
import * as https from "https";
import { URL } from "url";
import { securityHardeningSystem } from "./security-hardening-system.js";
import { frameworkLogger } from "../core/framework-logger.js";
/**
 * Secure authentication and authorization system
 */
export class SecureAuthenticationSystem extends EventEmitter {
    jwtSecret;
    jwtExpiresIn;
    refreshTokenExpiresIn;
    users = new Map();
    sessions = new Map();
    rbacConfig;
    oauth2Config;
    mfaEnabled = false;
    apiKeys = new Map();
    constructor(jwtSecret, rbacConfig, options = {}) {
        super();
        this.jwtSecret = jwtSecret;
        this.rbacConfig = rbacConfig;
        this.jwtExpiresIn = options.jwtExpiresIn || "1h";
        this.refreshTokenExpiresIn = options.refreshTokenExpiresIn || "7d";
        this.oauth2Config = options.oauth2Config || undefined;
        this.mfaEnabled = options.mfaEnabled || false;
        this.setupEventHandlers();
    }
    /**
     * Setup event handlers
     */
    setupEventHandlers() {
        this.on("authentication-success", this.handleAuthSuccess.bind(this));
        this.on("authentication-failure", this.handleAuthFailure.bind(this));
        this.on("authorization-denied", this.handleAuthzDenied.bind(this));
        this.on("session-created", this.handleSessionCreated.bind(this));
        this.on("session-destroyed", this.handleSessionDestroyed.bind(this));
    }
    /**
     * Authenticate user
     */
    async authenticate(request) {
        try {
            let user;
            switch (request.method) {
                case "password":
                    user = await this.authenticateWithPassword(request);
                    break;
                case "oauth2":
                    user = await this.authenticateWithOAuth2(request);
                    break;
                case "api_key":
                    user = await this.authenticateWithApiKey(request);
                    break;
                default:
                    throw new Error(`Unsupported authentication method: ${request.method}`);
            }
            if (!user) {
                this.emit("authentication-failure", {
                    method: request.method,
                    username: request.username,
                    email: request.email,
                    reason: "Invalid credentials",
                });
                return {
                    success: false,
                    error: "Invalid credentials",
                };
            }
            // Check if user is active
            if (!user.isActive) {
                this.emit("authentication-failure", {
                    method: request.method,
                    userId: user.id,
                    reason: "Account disabled",
                });
                return {
                    success: false,
                    error: "Account is disabled",
                };
            }
            // Check if MFA is required
            if (this.mfaEnabled && this.requiresMFA(user)) {
                return {
                    success: false,
                    requiresMFA: true,
                    user,
                };
            }
            // Generate tokens
            const token = this.generateJWT(user);
            const refreshToken = this.generateRefreshToken(user);
            // Update last login
            user.lastLoginAt = new Date();
            this.users.set(user.id, user);
            // Create session
            const sessionId = this.createSession(user, request.metadata);
            this.emit("authentication-success", {
                userId: user.id,
                method: request.method,
                sessionId,
            });
            return {
                success: true,
                user,
                token,
                refreshToken,
                expiresAt: this.getTokenExpiration(),
            };
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : "Authentication failed";
            this.emit("authentication-failure", {
                method: request.method,
                username: request.username,
                email: request.email,
                reason: errorMessage,
            });
            await frameworkLogger.log("secure-authentication-system", "-security-authentication-failed-errormessage-", "error", { message: `[SECURITY] Authentication failed: ${errorMessage}` });
            return {
                success: false,
                error: errorMessage,
            };
        }
    }
    /**
        * Authenticate with password
        */
    async authenticateWithPassword(request) {
        if (!request.username && !request.email) {
            throw new Error("Username or email is required");
        }
        if (!request.password) {
            throw new Error("Password is required for password authentication");
        }
        // Find user by username or email
        let user;
        for (const u of this.users.values()) {
            if ((request.username && u.username === request.username) ||
                (request.email && u.email === request.email)) {
                user = u;
                break;
            }
        }
        if (!user) {
            return undefined;
        }
        // Verify password using stored hash and salt
        if (!user.passwordHash || !user.passwordSalt) {
            return undefined;
        }
        const isValidPassword = await securityHardeningSystem.verifyPassword(request.password, user.passwordHash, user.passwordSalt);
        return isValidPassword ? user : undefined;
    }
    /**
        * Authenticate with OAuth2
        * SECURITY: Secure OAuth2 token validation with token endpoint validation
        */
    async authenticateWithOAuth2(request) {
        if (!this.oauth2Config) {
            throw new Error("OAuth2 not configured");
        }
        const { metadata, email } = request;
        if (!metadata || !metadata.code && !metadata.access_token) {
            throw new Error("OAuth2 requires code or access token in metadata");
        }
        try {
            let tokenResponse;
            // Handle different OAuth2 flows
            if (metadata.access_token) {
                // Direct token validation (e.g., from Single Page App)
                tokenResponse = {
                    access_token: metadata.access_token,
                    expires_in: metadata.expires_in || 3600,
                    token_type: metadata.token_type || "Bearer",
                };
            }
            else if (metadata.code) {
                // Authorization Code Grant Flow
                tokenResponse = await this.exchangeCodeForToken(metadata.code, metadata.redirect_uri);
            }
            else {
                throw new Error("Invalid OAuth2 flow");
            }
            // Validate token with provider's token endpoint
            const userInfo = await this.validateOAuth2Token(tokenResponse.access_token);
            // Find or create user based on OAuth2 email
            let user = this.findUserByEmail(userInfo.email);
            if (!user) {
                // Create new user from OAuth2 data
                user = await this.createUserFromOAuth2(userInfo, request.metadata || {});
            }
            else if (!user.isActive) {
                throw new Error("User account is disabled");
            }
            // Update OAuth2 metadata
            user.metadata = user.metadata || {};
            user.metadata.oauth2 = {
                provider: this.oauth2Config.clientId,
                lastLoginAt: new Date(),
                scopes: tokenResponse.scope,
                expiresAt: tokenResponse.expires_in ? new Date(Date.now() + tokenResponse.expires_in * 1000) : undefined,
            };
            this.users.set(user.id, user);
            return user;
        }
        catch (error) {
            const errorMessage = `OAuth2 authentication failed: ${error instanceof Error ? error.message : String(error)}`;
            frameworkLogger.log("secure-authentication-system", "-security-oauth2-authentication-failed-errormessage-", "error", { message: `[SECURITY] OAuth2 authentication failed: ${errorMessage}` });
            throw new Error(errorMessage);
        }
    }
    /**
     * Exchange OAuth2 authorization code for access token
     */
    async exchangeCodeForToken(code, redirectUri) {
        return new Promise((resolve, reject) => {
            const postData = JSON.stringify({
                grant_type: "authorization_code",
                client_id: this.oauth2Config.clientId,
                client_secret: this.oauth2Config.clientSecret,
                code: code,
                redirect_uri: redirectUri || this.oauth2Config.callbackURL,
            });
            const options = {
                hostname: new URL(this.oauth2Config.tokenURL).hostname,
                port: 443,
                path: new URL(this.oauth2Config.tokenURL).pathname,
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                    "Content-Length": Buffer.byteLength(postData),
                },
            };
            const req = https.request(options, (res) => {
                let data = "";
                res.on("data", (chunk) => { data += chunk; });
                res.on("end", () => {
                    if (res.statusCode === 200) {
                        try {
                            const tokenData = JSON.parse(data);
                            resolve({
                                access_token: tokenData.access_token,
                                refresh_token: tokenData.refresh_token,
                                expires_in: tokenData.expires_in,
                                token_type: tokenData.token_type,
                                scope: tokenData.scope,
                            });
                        }
                        catch (parseError) {
                            reject(new Error(`Invalid token response: ${parseError instanceof Error ? parseError.message : String(parseError)}`));
                        }
                    }
                    else {
                        reject(new Error(`Token exchange failed: ${res.statusCode} - ${data}`));
                    }
                });
            });
            req.on("error", (error) => {
                reject(new Error(`HTTP request failed: ${error.message}`));
            });
            req.write(postData);
            req.end();
        });
    }
    /**
     * Validate OAuth2 token with user info endpoint
     */
    async validateOAuth2Token(token) {
        return new Promise((resolve, reject) => {
            const options = {
                hostname: new URL(this.oauth2Config.tokenURL).hostname,
                port: 443,
                path: "/oauth/userinfo", // Common endpoint, adjust based on provider
                method: "GET",
                headers: {
                    "Authorization": `Bearer ${token}`,
                },
            };
            const req = https.request(options, (res) => {
                let data = "";
                res.on("data", (chunk) => { data += chunk; });
                res.on("end", () => {
                    if (res.statusCode === 200) {
                        try {
                            const userInfo = JSON.parse(data);
                            resolve({
                                id: userInfo.sub || userInfo.id,
                                email: userInfo.email,
                                name: userInfo.name || userInfo.display_name,
                                picture: userInfo.picture,
                            });
                        }
                        catch (parseError) {
                            reject(new Error(`Invalid user info response: ${parseError instanceof Error ? parseError.message : String(parseError)}`));
                        }
                    }
                    else {
                        reject(new Error(`Token validation failed: ${res.statusCode} - ${data}`));
                    }
                });
            });
            req.on("error", (error) => {
                reject(new Error(`HTTP request failed: ${error.message}`));
            });
            req.end();
        });
    }
    /**
     * Find user by email
     */
    findUserByEmail(email) {
        for (const user of this.users.values()) {
            if (user.email === email) {
                return user;
            }
        }
        return undefined;
    }
    /**
     * Create user from OAuth2 data
     */
    async createUserFromOAuth2(userInfo, metadata) {
        const userId = crypto.randomUUID();
        const username = userInfo.name || userInfo.email.split('@')[0];
        // Create base user with OAuth2 roles
        const user = {
            id: userId,
            username: `${username}_${Date.now()}`, // Ensure uniqueness
            email: userInfo.email,
            roles: ["user"], // Default role
            permissions: ["authenticated"],
            metadata: {
                ...metadata,
                oauth2: {
                    provider: this.oauth2Config.clientId,
                    createdAt: new Date(),
                    scopes: metadata.scopes || [],
                },
            },
            createdAt: new Date(),
            isActive: true,
        };
        this.users.set(userId, user);
        return user;
    }
    /**
        * Authenticate with API key
        * SECURITY: Secure API key validation with rate limiting and usage tracking
        */
    async authenticateWithApiKey(request) {
        if (!request.metadata || !request.metadata.api_key) {
            throw new Error("API key is required in metadata");
        }
        const apiKeyValue = request.metadata.api_key;
        // Find API key and validate
        const apiKey = Array.from(this.apiKeys.values()).find(key => key.key === apiKeyValue && key.isActive);
        if (!apiKey) {
            throw new Error("Invalid or inactive API key");
        }
        // Check if API key is expired
        if (apiKey.expiresAt && new Date() > apiKey.expiresAt) {
            throw new Error("API key expired");
        }
        // Get user associated with API key
        const user = this.users.get(apiKey.userId);
        if (!user) {
            throw new Error("User associated with API key not found");
        }
        if (!user.isActive) {
            throw new Error("User account is disabled");
        }
        // Update API key usage tracking
        apiKey.lastUsedAt = new Date();
        this.apiKeys.set(apiKey.id, apiKey);
        // Update user metadata
        user.metadata = user.metadata || {};
        user.metadata.apiKeyUsage = {
            lastUsedAt: new Date(),
            apiKeyId: apiKey.id,
            apiKeyName: apiKey.name,
        };
        return user;
    }
    /**
     * Authorize user action
     */
    authorize(request) {
        try {
            const { user, resource, action } = request;
            // Check if user has required permissions
            const requiredPermissions = this.getRequiredPermissions(resource, action);
            const userPermissions = this.getUserPermissions(user);
            const hasPermission = requiredPermissions.every((perm) => userPermissions.includes(perm));
            if (!hasPermission) {
                const missingPermissions = requiredPermissions.filter((perm) => !userPermissions.includes(perm));
                this.emit("authorization-denied", {
                    userId: user.id,
                    resource,
                    action,
                    requiredPermissions,
                    missingPermissions,
                });
                return {
                    allowed: false,
                    reason: "Insufficient permissions",
                    requiredPermissions,
                    missingPermissions,
                };
            }
            return {
                allowed: true,
            };
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : "Authorization failed";
            frameworkLogger.log("secure-authentication-system", "-security-authorization-failed-errormessage-", "error", { message: `[SECURITY] Authorization failed: ${errorMessage}` });
            return {
                allowed: false,
                reason: errorMessage,
            };
        }
    }
    /**
     * Verify JWT token
     */
    verifyToken(token) {
        try {
            const payload = jwt.verify(token, this.jwtSecret);
            // Additional validation
            if (payload.exp && payload.exp < Date.now() / 1000) {
                return { valid: false, error: "Token expired" };
            }
            return { valid: true, payload };
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : "Token verification failed";
            return { valid: false, error: errorMessage };
        }
    }
    /**
     * Refresh access token
     */
    refreshToken(refreshToken) {
        try {
            // Verify refresh token (simplified - in production use proper refresh token validation)
            const payload = jwt.verify(refreshToken, this.jwtSecret);
            // Find user
            const user = this.users.get(payload.userId);
            if (!user) {
                return { success: false, error: "User not found" };
            }
            // Generate new token
            const newToken = this.generateJWT(user);
            return { success: true, token: newToken };
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : "Token refresh failed";
            return { success: false, error: errorMessage };
        }
    }
    /**
     * Create user session
     */
    createSession(user, metadata) {
        const sessionId = securityHardeningSystem.generateSecureToken(32);
        const expiresAt = new Date(Date.now() + 24 * 60 * 60 * 1000); // 24 hours
        this.sessions.set(sessionId, {
            userId: user.id,
            expiresAt,
            metadata: metadata || {},
        });
        this.emit("session-created", { sessionId, userId: user.id });
        return sessionId;
    }
    /**
     * Destroy user session
     */
    destroySession(sessionId) {
        const session = this.sessions.get(sessionId);
        if (session) {
            this.sessions.delete(sessionId);
            this.emit("session-destroyed", { sessionId, userId: session.userId });
            return true;
        }
        return false;
    }
    /**
     * Validate session
     */
    validateSession(sessionId) {
        const session = this.sessions.get(sessionId);
        if (!session) {
            return { valid: false, error: "Session not found" };
        }
        if (session.expiresAt < new Date()) {
            this.sessions.delete(sessionId);
            return { valid: false, error: "Session expired" };
        }
        const user = this.users.get(session.userId);
        if (!user) {
            return { valid: false, error: "User not found" };
        }
        return { valid: true, user };
    }
    /**
     * Create new user
     */
    async createUser(userData) {
        // Validate input
        const validation = securityHardeningSystem.validateInput(userData, "user-creation");
        if (!validation.isValid) {
            throw new Error(`Invalid user data: ${validation.errors.join(", ")}`);
        }
        // Check if user already exists
        for (const user of this.users.values()) {
            if (user.username === userData.username ||
                user.email === userData.email) {
                throw new Error("User already exists");
            }
        }
        // Hash password
        const passwordResult = await securityHardeningSystem.hashPassword(userData.password);
        // Create user
        const user = {
            id: securityHardeningSystem.generateSecureToken(16),
            username: userData.username,
            email: userData.email,
            roles: userData.roles || ["user"],
            permissions: this.getPermissionsForRoles(userData.roles || ["user"]),
            metadata: userData.metadata || {},
            createdAt: new Date(),
            isActive: true,
            passwordHash: passwordResult.hash,
            passwordSalt: passwordResult.salt,
        };
        // Store user (in production, this would be in a database)
        this.users.set(user.id, user);
        // Store password hash separately (in production, this would be in a secure database)
        // For demo purposes, we'll skip this
        return user;
    }
    /**
     * Get required permissions for resource and action
     */
    getRequiredPermissions(resource, action) {
        // This would be configured based on your RBAC system
        // For demo purposes, return basic permissions
        return [`${resource}:${action}`];
    }
    /**
     * Get user permissions including inherited permissions from roles
     */
    getUserPermissions(user) {
        const permissions = new Set(user.permissions);
        // Add permissions from roles
        user.roles.forEach((role) => {
            const rolePermissions = this.rbacConfig.permissions[role] || [];
            rolePermissions.forEach((perm) => permissions.add(perm));
            // Add permissions from parent roles
            const parentRoles = this.rbacConfig.roleHierarchy[role] || [];
            parentRoles.forEach((parentRole) => {
                const parentPermissions = this.rbacConfig.permissions[parentRole] || [];
                parentPermissions.forEach((perm) => permissions.add(perm));
            });
        });
        return Array.from(permissions);
    }
    /**
     * Get permissions for roles
     */
    getPermissionsForRoles(roles) {
        const permissions = new Set();
        roles.forEach((role) => {
            const rolePermissions = this.rbacConfig.permissions[role] || [];
            rolePermissions.forEach((perm) => permissions.add(perm));
            // Add permissions from parent roles
            const parentRoles = this.rbacConfig.roleHierarchy[role] || [];
            parentRoles.forEach((parentRole) => {
                const parentPermissions = this.rbacConfig.permissions[parentRole] || [];
                parentPermissions.forEach((perm) => permissions.add(perm));
            });
        });
        return Array.from(permissions);
    }
    /**
     * Generate JWT token
     */
    generateJWT(user) {
        const payload = {
            userId: user.id,
            username: user.username,
            email: user.email,
            roles: user.roles,
            permissions: Array.from(new Set(user.permissions)),
            iss: "strray-framework",
            aud: "strray-api",
        };
        return jwt.sign(payload, this.jwtSecret, {
            expiresIn: "1h",
        });
    }
    /**
     * Generate refresh token
     */
    generateRefreshToken(user) {
        const payload = {
            userId: user.id,
            type: "refresh",
            iss: "strray-framework",
            aud: "strray-api",
        };
        return jwt.sign(payload, this.jwtSecret, {
            expiresIn: "7d",
        });
    }
    /**
     * Get token expiration date
     */
    getTokenExpiration() {
        const now = new Date();
        // For '1h', add 1 hour
        now.setHours(now.getHours() + 1);
        return now;
    }
    /**
     * Check if user requires MFA
     */
    requiresMFA(user) {
        // Implement MFA requirement logic based on user roles, risk level, etc.
        return user.roles.includes("admin") || user.roles.includes("manager");
    }
    /**
     * Verify password (placeholder - in production, verify against stored hash)
     */
    async verifyPassword(password, userId) {
        // In production, retrieve stored hash and verify
        // For demo purposes, return true
        return true;
    }
    /**
     * Event handlers
     */
    handleAuthSuccess(event) {
        frameworkLogger.log("secure-authentication-system", "-authentication-successful-for-user-event-userid-", "success", { message: `✅ Authentication successful for user: ${event.userId}` });
    }
    handleAuthFailure(event) {
        console.warn(`❌ Authentication failed: ${event.reason}`);
    }
    handleAuthzDenied(event) {
        console.warn(`🚫 Authorization denied for user ${event.userId}: ${event.resource}:${event.action}`);
    }
    handleSessionCreated(event) {
        frameworkLogger.log("secure-authentication-system", "-session-created-event-sessionid-for-user-event-us", "info", {
            message: `📋 Session created: ${event.sessionId} for user ${event.userId}`,
        });
    }
    handleSessionDestroyed(event) {
        frameworkLogger.log("secure-authentication-system", "-session-destroyed-event-sessionid-for-user-event-", "info", {
            message: `📋 Session destroyed: ${event.sessionId} for user ${event.userId}`,
        });
    }
    /**
     * Get user by ID
     */
    getUser(userId) {
        return this.users.get(userId);
    }
    /**
     * Get all users
     */
    getAllUsers() {
        return Array.from(this.users.values());
    }
    /**
     * Update user
     */
    updateUser(userId, updates) {
        const user = this.users.get(userId);
        if (!user)
            return false;
        Object.assign(user, updates);
        this.users.set(userId, user);
        return true;
    }
    /**
     * Delete user
     */
    deleteUser(userId) {
        return this.users.delete(userId);
    }
    /**
       * Create new API key for user
       * SECURITY: Generate cryptographically secure API keys with usage tracking
       */
    createAPIKey(userId, options) {
        // Validate user exists
        const user = this.users.get(userId);
        if (!user) {
            throw new Error("User not found");
        }
        // Generate secure API key
        const apiKey = this.generateSecureAPIKey();
        // Create API key object
        const newApiKey = {
            id: securityHardeningSystem.generateSecureToken(16),
            key: apiKey,
            userId,
            name: options.name,
            permissions: options.permissions || user.permissions,
            lastUsedAt: options.expiresAt instanceof Date ? new Date() : undefined,
            createdAt: new Date(),
            expiresAt: options.expiresAt instanceof Date ? options.expiresAt : undefined,
            isActive: true,
            rateLimit: options.rateLimit,
        };
        // Store API key
        this.apiKeys.set(newApiKey.id, newApiKey);
        // Log API key creation
        frameworkLogger.log("secure-authentication-system", "-api-key-created-event-api-key-id-for-user-event-userid-", "info", {
            message: `🔑 API key created: ${newApiKey.id} for user ${userId}`,
            apiKeyId: newApiKey.id,
            userId,
            permissions: newApiKey.permissions,
            expiresAt: newApiKey.expiresAt,
        });
        return newApiKey;
    }
    /**
     * Revoke API key by ID
     */
    revokeAPIKey(apiKeyId, userId) {
        const apiKey = this.apiKeys.get(apiKeyId);
        if (!apiKey) {
            throw new Error("API key not found");
        }
        if (apiKey.userId !== userId) {
            throw new Error("Unauthorized to revoke this API key");
        }
        // Deactivate API key
        apiKey.isActive = false;
        this.apiKeys.set(apiKeyId, apiKey);
        // Log API key revocation
        frameworkLogger.log("secure-authentication-system", "-api-key-revoked-event-api-key-id-for-user-event-userid-", "info", {
            message: `🔓 API key revoked: ${apiKeyId} for user ${userId}`,
            apiKeyId,
            userId,
        });
        return true;
    }
    /**
     * Get API keys for a user
     */
    getAPIKeys(userId) {
        const userKeys = Array.from(this.apiKeys.values())
            .filter(key => key.userId === userId)
            .map(key => ({
            ...key,
            key: this.maskAPIKey(key.key), // Mask the key for security
        }));
        return userKeys;
    }
    /**
     * Get API key by ID (admin only)
     */
    getAPIKeyById(apiKeyId, requestingUserId) {
        const apiKey = this.apiKeys.get(apiKeyId);
        if (!apiKey) {
            return undefined;
        }
        // Check if user owns this API key or has admin permissions
        if (requestingUserId && apiKey.userId !== requestingUserId) {
            const user = this.users.get(requestingUserId);
            if (!user || !user.permissions.includes("admin")) {
                throw new Error("Unauthorized to access this API key");
            }
        }
        return {
            ...apiKey,
            key: this.maskAPIKey(apiKey.key), // Mask the key for security
        };
    }
    /**
     * Update API key properties
     */
    updateAPIKey(apiKeyId, userId, updates) {
        const apiKey = this.apiKeys.get(apiKeyId);
        if (!apiKey) {
            throw new Error("API key not found");
        }
        if (apiKey.userId !== userId) {
            throw new Error("Unauthorized to update this API key");
        }
        // Update API key properties
        const updatedApiKey = {
            ...apiKey,
            ...updates,
            updatedAt: new Date(), // Add update timestamp
        };
        this.apiKeys.set(apiKeyId, updatedApiKey);
        // Log API key update
        frameworkLogger.log("secure-authentication-system", "-api-key-updated-event-api-key-id-for-user-event-userid-", "info", {
            message: `📝 API key updated: ${apiKeyId} for user ${userId}`,
            apiKeyId,
            userId,
            updatedFields: Object.keys(updates),
        });
        return true;
    }
    /**
     * Check API key rate limit
     */
    async checkRateLimit(apiKey) {
        // Find API key
        const apiKeyData = Array.from(this.apiKeys.values()).find(key => key.key === apiKey);
        if (!apiKeyData || !apiKeyData.isActive) {
            return { allowed: false, remaining: 0, resetAt: new Date() };
        }
        // Check expiration
        if (apiKeyData.expiresAt && new Date() > apiKeyData.expiresAt) {
            return { allowed: false, remaining: 0, resetAt: new Date() };
        }
        // Get or create rate limit tracking
        // In production, this would be persisted to a database
        const rateLimit = apiKeyData.rateLimit;
        if (!rateLimit) {
            return { allowed: true, remaining: -1, resetAt: new Date() }; // No rate limit
        }
        // Simple in-memory rate limiting (production would use Redis/database)
        // For demo purposes, we'll accept all requests with rate limits configured
        return {
            allowed: true,
            remaining: rateLimit.requests,
            resetAt: new Date(Date.now() + rateLimit.window * 60 * 1000)
        };
    }
    /**
     * Rotate API keys (security best practice)
     */
    async rotateAPIKeys(userId) {
        // Get user's active API keys
        const userKeys = Array.from(this.apiKeys.values())
            .filter(key => key.userId === userId && key.isActive);
        if (userKeys.length === 0) {
            throw new Error("No active API keys to rotate");
        }
        // Revoke the oldest active API key
        const apiKeyToRevoke = userKeys.reduce((oldest, current) => current.createdAt < oldest.createdAt ? current : oldest);
        // Create new API key with same properties
        const newApiKey = this.createAPIKey(userId, {
            name: `${apiKeyToRevoke.name} (rotated)`,
            permissions: apiKeyToRevoke.permissions,
            expiresAt: apiKeyToRevoke.expiresAt instanceof Date ? apiKeyToRevoke.expiresAt : undefined,
            rateLimit: apiKeyToRevoke.rateLimit,
        });
        // Revoke the old API key
        this.revokeAPIKey(apiKeyToRevoke.id, userId);
        return { newApiKey, revokedApiKey: apiKeyToRevoke };
    }
    /**
     * Get authentication statistics
     */
    getAuthStats() {
        const activeUsers = Array.from(this.users.values()).filter((u) => u.isActive).length;
        const totalAPIKeys = this.apiKeys.size;
        const activeAPIKeys = Array.from(this.apiKeys.values()).filter((key) => key.isActive).length;
        return {
            totalUsers: this.users.size,
            activeUsers,
            totalSessions: this.sessions.size,
            totalAPIKeys,
            activeAPIKeys,
            recentEvents: [], // Would track recent auth events
        };
    }
    /**
     * Generate cryptographically secure API key
     */
    generateSecureAPIKey() {
        // Generate a secure token for API key
        const prefix = "sk_" + crypto
            .randomBytes(32)
            .toString('base64')
            .replace(/\+/g, '-')
            .replace(/\//g, '_')
            .replace(/=/g, '');
        return prefix;
    }
    /**
     * Mask API key for display purposes (security)
     */
    maskAPIKey(apiKey) {
        if (apiKey.length <= 8) {
            return '*'.repeat(apiKey.length);
        }
        const prefix = apiKey.substring(0, 4);
        const suffix = apiKey.substring(apiKey.length - 4);
        const masked = '*'.repeat(apiKey.length - 8);
        return `${prefix}${masked}${suffix}`;
    }
}
// Export singleton instance (would need proper initialization)
export const secureAuthSystem = {}; // Placeholder
