// strray-ai/dist/core/framework-logger.js
import { isLoggingEnabled, shouldLog, } from "./logging-config.js";
/**
 * Generate a unique job ID for tracking work sessions
 */
export function generateJobId(prefix = "job") {
    const timestamp = Date.now();
    const random = Math.random().toString(36).substring(2, 8);
    return `${prefix}-${timestamp}-${random}`;
}
// Global job context for correlation across all framework operations
let currentJobContext = null;
export function getCurrentJobId() {
    return currentJobContext?.jobId || null;
}
export function setCurrentJobContext(jobId) {
    currentJobContext = new JobContext(jobId);
    return currentJobContext;
}
export function withJobContext(operation, jobId) {
    const originalContext = currentJobContext;
    const jobContext = setCurrentJobContext(jobId);
    try {
        const result = operation();
        if (result instanceof Promise) {
            return result.finally(async () => {
                // Auto-complete job on operation finish
                await jobContext.complete(true).catch(console.error);
                // Restore original context
                currentJobContext = originalContext;
            });
        }
        else {
            // Sync operation - complete immediately
            jobContext.complete(true).catch(console.error);
            currentJobContext = originalContext;
            return Promise.resolve(result);
        }
    }
    catch (error) {
        // Error occurred - complete job with failure
        jobContext.complete(false, { error: String(error) }).catch(console.error);
        currentJobContext = originalContext;
        throw error;
    }
}
export class JobContext {
    jobId;
    startTime;
    complexityScore;
    agentUsed;
    operationType;
    // NEW: Outcome tracking for pattern analysis
    outcome;
    // NEW: Predicted duration for accuracy comparison (in ms)
    predictedDuration;
    constructor(jobId) {
        this.jobId = jobId || generateJobId("auto");
        this.startTime = Date.now();
    }
    /**
     * Set the outcome after task completion
     */
    setOutcome(success, escalated = false, autoFixed = false) {
        if (escalated)
            this.outcome = "escalated";
        else if (autoFixed)
            this.outcome = "auto-fixed";
        else
            this.outcome = success ? "success" : "fail";
    }
    /**
     * Set predicted duration for complexity accuracy comparison
     */
    setPredictedDuration(completionTimeMs) {
        this.predictedDuration = completionTimeMs;
    }
    /**
     * Log job completion with diagnostic info
     * Enhanced with outcome and complexity accuracy tracking
     */
    async complete(success = true, details) {
        const actualDuration = Date.now() - this.startTime;
        // Calculate complexity accuracy if we have both predicted and actual
        let complexityAccuracy;
        if (this.complexityScore && this.predictedDuration) {
            const ratio = actualDuration / this.predictedDuration;
            if (ratio > 1.5) {
                complexityAccuracy = "underestimated";
            }
            else if (ratio < 0.5) {
                complexityAccuracy = "overestimated";
            }
            else {
                complexityAccuracy = "accurate";
            }
        }
        else if (this.complexityScore && actualDuration) {
            // Fallback: estimate predicted based on complexity score
            const estimatedPredicted = this.complexityScore * 1000; // 1 second per complexity point
            const ratio = actualDuration / estimatedPredicted;
            if (ratio > 1.5) {
                complexityAccuracy = "underestimated";
            }
            else if (ratio < 0.5) {
                complexityAccuracy = "overestimated";
            }
            else {
                complexityAccuracy = "accurate";
            }
        }
        // Determine final outcome
        const finalOutcome = this.outcome || (success ? "success" : "fail");
        await frameworkLogger.log("job-context", "job-completed", success ? "success" : "error", {
            duration: actualDuration,
            complexityScore: this.complexityScore,
            agentUsed: this.agentUsed,
            operationType: this.operationType,
            outcome: finalOutcome,
            complexityAccuracy,
            predictedDuration: this.predictedDuration,
            ...details,
        }, undefined, // sessionId
        this.jobId);
    }
}
export class FrameworkUsageLogger {
    logs = [];
    maxLogs = 1000;
    async log(component, action, status, details, sessionId, jobId) {
        // Check if logging is enabled globally
        if (!isLoggingEnabled()) {
            return;
        }
        // Check if this log level should be logged
        if (!shouldLog(status)) {
            return;
        }
        // Use current job context if available, otherwise auto-generate
        const actualJobId = jobId || getCurrentJobId() || generateJobId("auto");
        // Ensure we always have a jobId
        if (!actualJobId) {
            throw new Error("JobId generation failed");
        }
        const entry = {
            timestamp: Date.now(),
            component,
            action,
            status,
            agent: "orchestrator",
            ...(sessionId && { sessionId }),
            jobId: actualJobId,
            details,
        };
        this.logs.push(entry);
        if (this.logs.length > this.maxLogs) {
            this.logs.shift();
        }
        // Always persist to file, never output to console to avoid UI bleed-through
        try {
            this.persistLog(entry);
        }
        catch (error) {
            // Silently fail - logging should never break application
        }
    }
    async persistLog(entry) {
        // Write to log file for monitoring subagent
        try {
            const fs = await import("fs");
            const path = await import("path");
            const cwd = process.cwd();
            if (!cwd) {
                // Skip logging if cwd is not available (e.g., in test environments)
                return;
            }
            const logDir = path.join(cwd, "logs", "framework");
            const logFile = path.join(logDir, "activity.log");
            // Ensure log directory exists
            if (!fs.existsSync(logDir)) {
                fs.mkdirSync(logDir, { recursive: true });
            }
            const jobIdPart = entry.jobId ? `[${entry.jobId}] ` : "";
            const logEntry = `${new Date(entry.timestamp).toISOString()} ${jobIdPart}[${entry.component}] ${entry.action} - ${entry.status.toUpperCase()}\n`;
            fs.appendFileSync(logFile, logEntry);
        }
        catch (error) {
            // Silent fail - cannot log to console as this IS the logger
        }
    }
    getRecentLogs(count = 50) {
        return this.logs.slice(-count);
    }
    getComponentUsage(component) {
        return this.logs.filter((log) => log.component === component);
    }
    printRundown() {
        // Framework usage analytics - available for debugging but should not output to console
        // Use getRecentLogs() directly instead of printing
    }
}
export const frameworkLogger = new FrameworkUsageLogger();
