// strray-ai/dist/core/model-router.js
/**
 * Model Router
 *
 * Intelligent model routing based on task complexity and type.
 * Integrates with features-config for task-based model selection.
 *
 * @version 2.0.0
 * @since 2026-01-25
 */
import * as fs from "fs";
import * as path from "path";
import { featuresConfigLoader, detectTaskType, } from "./features-config.js";
// ============================================================================
// Model Router Class
// ============================================================================
class ModelRouter {
    configPath;
    config;
    availableModels;
    constructor(configPath) {
        this.configPath =
            configPath || path.resolve(process.cwd(), "opencode.json");
        this.config = this.loadConfig();
        this.availableModels = this.discoverModels();
    }
    loadConfig() {
        try {
            const configData = fs.readFileSync(this.configPath, "utf8");
            return JSON.parse(configData);
        }
        catch (error) {
            // Provide defaults if config not found
            return {
                model_routing: {},
                model_default: "claude-sonnet-4",
                model_fallback: "claude-haiku-4",
            };
        }
    }
    discoverModels() {
        const defaultModels = [
            // Anthropic Claude models (preferred for 0xRay)
            "claude-opus-4",
            "claude-sonnet-4",
            "claude-haiku-4",
            // Other supported models
            "openai/gpt-4o",
            "openai/gpt-4o-mini",
            "google/gemini-pro-1.5",
            "google/gemini-flash-1.5",
            "meta/llama-3.1-70b-instruct",
            "mistral/mistral-7b-instruct",
        ];
        const configModels = this.config.available_models || [];
        const uniqueModels = new Set([...defaultModels, ...configModels]);
        return Array.from(uniqueModels);
    }
    /**
     * Get the appropriate model for a specific task context
     * Uses features-config for task-based routing when available
     */
    getModelForTask(context) {
        const featuresConfig = featuresConfigLoader.loadConfig();
        if (featuresConfig.model_routing.enabled && context.toolName) {
            const taskContext = {};
            if (context.fileCount !== undefined)
                taskContext.fileCount = context.fileCount;
            if (context.isComplex !== undefined)
                taskContext.isComplex = context.isComplex;
            const taskType = detectTaskType(context.toolName, taskContext);
            if (taskType !== "unknown") {
                const model = featuresConfigLoader.getModelForTask(taskType);
                if (this.isModelAvailable(model)) {
                    return model;
                }
            }
        }
        // Fall back to agent-specific routing
        if (context.agentType) {
            return this.getValidatedModel(context.agentType);
        }
        return this.getValidatedModel();
    }
    /**
     * Get validated model for agent type (legacy method)
     */
    getValidatedModel(agentType) {
        // 1. User preference (future enhancement)
        const userModel = this.getUserPreference();
        if (userModel && this.isModelAvailable(userModel)) {
            return userModel;
        }
        // 2. Check features config for agent-specific model
        if (agentType) {
            const featuresConfig = featuresConfigLoader.loadConfig();
            const agentModel = featuresConfig.agent_management.agent_models[agentType];
            if (agentModel && this.isModelAvailable(agentModel)) {
                return agentModel;
            }
            // Fall back to opencode.json model routing
            if (this.config.model_routing?.[agentType]) {
                const configModel = this.config.model_routing[agentType];
                if (this.isModelAvailable(configModel)) {
                    return configModel;
                }
            }
        }
        // 3. Framework default from features config
        const featuresConfig = featuresConfigLoader.loadConfig();
        const defaultModel = featuresConfig.model_routing?.default_model;
        if (defaultModel && this.isModelAvailable(defaultModel)) {
            return defaultModel;
        }
        // 4. Legacy default
        const legacyDefault = this.config.model_default || "claude-sonnet-4";
        if (this.isModelAvailable(legacyDefault)) {
            return legacyDefault;
        }
        // 5. Ultimate fallback
        const fallbackModel = featuresConfig.model_routing?.fallback_model ||
            this.config.model_fallback ||
            "claude-haiku-4";
        return fallbackModel;
    }
    /**
     * Get model for a specific task type directly
     */
    getModelForTaskType(taskType) {
        return featuresConfigLoader.getModelForTask(taskType);
    }
    /**
     * Get max tokens for a task type
     */
    getMaxTokensForTaskType(taskType) {
        return featuresConfigLoader.getMaxTokensForTask(taskType);
    }
    getUserPreference() {
        // Check for user-specific config (future enhancement)
        return null;
    }
    isModelAvailable(model) {
        if (this.config.deprecated_models?.includes(model)) {
            return false;
        }
        // For now, accept all models - the API will validate
        return true;
    }
    /**
     * Reload configuration from disk
     */
    reloadConfig() {
        this.config = this.loadConfig();
        this.availableModels = this.discoverModels();
        featuresConfigLoader.clearCache();
    }
    /**
     * Get current configuration summary
     */
    getConfigSummary() {
        const featuresConfig = featuresConfigLoader.loadConfig();
        return {
            defaultModel: featuresConfig.model_routing?.default_model || "claude-sonnet-4",
            fallbackModel: featuresConfig.model_routing?.fallback_model || "claude-haiku-4",
            taskRoutingEnabled: featuresConfig.model_routing?.enabled || false,
            availableModels: this.availableModels,
        };
    }
}
// Export singleton instance
export const modelRouter = new ModelRouter();
export default modelRouter;
// Export convenience functions
export const getModelForTask = (context) => modelRouter.getModelForTask(context);
export const getModelForTaskType = (taskType) => modelRouter.getModelForTaskType(taskType);
export const getValidatedModel = (agentType) => modelRouter.getValidatedModel(agentType);
