// strray-ai/dist/mcps/mcp-client.js
/**
 * MCP Client Layer
 *
 * Enables framework components to call MCP servers programmatically.
 * Refactored as a lean facade using extracted modules (Phases 1-5).
 *
 * Architecture:
 * - ToolRegistry: Manages tool registration and lookup
 * - ToolDiscovery: Discovers tools from MCP servers
 * - ToolExecutor: Executes tools via JSON-RPC
 * - ToolCache: Caches discovered tools
 * - SimulationEngine: Provides fallback simulations
 */
import { frameworkLogger } from '../core/framework-logger.js';
import { defaultServerRegistry } from './config/index.js';
import { ToolRegistry, ToolDiscovery, ToolExecutor, ToolCache, } from './tools/index.js';
import { SimulationEngine, getAllServerSimulations, } from './simulation/index.js';
/**
 * MCP Client
 *
 * Facade that orchestrates tool discovery, caching, execution, and simulation.
 */
export class MCPClient {
    config;
    toolRegistry;
    toolDiscovery;
    toolExecutor;
    toolCache;
    simulationEngine;
    constructor(config) {
        this.config = config;
        this.toolRegistry = new ToolRegistry();
        this.toolDiscovery = new ToolDiscovery();
        this.toolExecutor = new ToolExecutor();
        this.toolCache = new ToolCache();
        this.simulationEngine = new SimulationEngine();
        this.registerDefaultSimulations();
    }
    /**
     * Register default simulation implementations
     */
    registerDefaultSimulations() {
        const simulations = getAllServerSimulations();
        for (const [serverName, serverSimulations] of Object.entries(simulations)) {
            this.simulationEngine.registerServerSimulators(serverName, serverSimulations);
        }
    }
    /**
     * Initialize MCP client by discovering and caching tools
     */
    async initialize() {
        const jobId = `mcp-init-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
        try {
            frameworkLogger.log('mcp-client', `initializing MCP client for ${this.config.serverName}`, 'info', { jobId });
            // Discover tools (currently uses simulation/static discovery)
            await this.discoverTools();
            frameworkLogger.log('mcp-client', `MCP client initialized with ${this.toolRegistry.getToolCount()} tools`, 'success', { jobId });
        }
        catch (error) {
            frameworkLogger.log('mcp-client', `failed to initialize MCP client: ${error instanceof Error ? error.message : String(error)}`, 'error', { jobId, error });
            throw error;
        }
    }
    /**
     * Discover available tools for this server
     */
    async discoverTools() {
        // Check cache first
        const cachedTools = this.toolCache.get(this.config.serverName);
        if (cachedTools) {
            this.toolRegistry.register(this.config.serverName, cachedTools);
            return;
        }
        // Static tool definitions by server (simulated discovery)
        const tools = this.getStaticTools(this.config.serverName);
        this.toolRegistry.register(this.config.serverName, tools);
        this.toolCache.set(this.config.serverName, tools);
    }
    /**
     * Get static tool definitions for a server
     */
    getStaticTools(serverName) {
        const staticTools = {
            'code-review': [
                {
                    name: 'analyze_code_quality',
                    description: 'Analyze code for quality, patterns, and best practices',
                    inputSchema: {
                        type: 'object',
                        properties: {
                            code: { type: 'string' },
                            language: { type: 'string' },
                            context: { type: 'object' },
                        },
                        required: ['code'],
                    },
                },
            ],
            'security-audit': [
                {
                    name: 'scan_vulnerabilities',
                    description: 'Scan code for security vulnerabilities and compliance issues',
                    inputSchema: {
                        type: 'object',
                        properties: {
                            files: { type: 'array', items: { type: 'string' } },
                            severity: {
                                type: 'string',
                                enum: ['low', 'medium', 'high', 'critical'],
                            },
                        },
                        required: ['files'],
                    },
                },
            ],
            'performance-optimization': [
                {
                    name: 'analyze_performance',
                    description: 'Analyze code for performance bottlenecks and optimization opportunities',
                    inputSchema: {
                        type: 'object',
                        properties: {
                            code: { type: 'string' },
                            language: { type: 'string' },
                            metrics: { type: 'array', items: { type: 'string' } },
                        },
                        required: ['code'],
                    },
                },
            ],
            'testing-strategy': [
                {
                    name: 'analyze_test_coverage',
                    description: 'Analyze test coverage and suggest testing strategies',
                    inputSchema: {
                        type: 'object',
                        properties: {
                            code: { type: 'string' },
                            existingTests: { type: 'array', items: { type: 'string' } },
                            requirements: { type: 'object' },
                        },
                        required: ['code'],
                    },
                },
            ],
            'researcher': [
                {
                    name: 'analyze_codebase',
                    description: 'Analyze complete codebase structure and provide insights',
                    inputSchema: {
                        type: 'object',
                        properties: {
                            scope: { type: 'string', enum: ['full', 'directory', 'file'] },
                            analysis: { type: 'array', items: { type: 'string' } },
                        },
                    },
                },
            ],
            'framework-help': [
                {
                    name: 'strray_get_capabilities',
                    description: 'Get StringRay framework capabilities',
                    inputSchema: { type: 'object', properties: {} },
                },
                {
                    name: 'strray_get_commands',
                    description: 'Get available StringRay commands',
                    inputSchema: { type: 'object', properties: {} },
                },
                {
                    name: 'strray_explain_capability',
                    description: 'Explain a specific capability',
                    inputSchema: {
                        type: 'object',
                        properties: {
                            capability: { type: 'string' },
                        },
                        required: ['capability'],
                    },
                },
            ],
            'skill-invocation': [
                {
                    name: 'invoke-skill',
                    description: 'Generic skill invocation tool for calling any MCP skill server',
                    inputSchema: {
                        type: 'object',
                        properties: {
                            skillName: { type: 'string' },
                            toolName: { type: 'string' },
                            args: { type: 'object' },
                        },
                        required: ['skillName', 'toolName'],
                    },
                },
            ],
            'strategist': [
                {
                    name: 'strategic_guidance',
                    description: 'Strategic guidance and complex problem-solving',
                    inputSchema: {
                        type: 'object',
                        properties: {
                            question: { type: 'string' },
                            context: { type: 'object' },
                            scope: { type: 'string', enum: ['technical', 'business', 'strategic'] },
                        },
                        required: ['question'],
                    },
                },
            ],
        };
        return staticTools[serverName] || [];
    }
    /**
     * Call a specific MCP server tool
     */
    async callTool(toolName, args = {}) {
        // Try simulation first (fallback behavior)
        if (this.simulationEngine.canSimulate(this.config.serverName, toolName)) {
            try {
                return await this.simulationEngine.simulate(this.config.serverName, toolName, args);
            }
            catch (error) {
                frameworkLogger.log('mcp-client', `Simulation failed for ${toolName}: ${error instanceof Error ? error.message : String(error)}`, 'info', { toolName });
            }
        }
        // Return generic fallback result
        return {
            content: [
                {
                    type: 'text',
                    text: `Tool ${toolName} executed on ${this.config.serverName} server`,
                },
            ],
        };
    }
    /**
     * Get list of available tools
     */
    getAvailableTools() {
        return this.toolRegistry.getTools(this.config.serverName);
    }
    /**
     * Check if a tool is available
     */
    hasTool(toolName) {
        return this.toolRegistry.hasTool(this.config.serverName, toolName);
    }
    /**
     * Get tool by name
     */
    getTool(toolName) {
        return this.toolRegistry.getTool(this.config.serverName, toolName);
    }
}
/**
 * MCP Client Manager
 *
 * Manages MCP client instances and provides unified interface
 * for framework components to access MCP server capabilities.
 */
export class MCPClientManager {
    static instance;
    clients = new Map();
    constructor() { }
    static getInstance() {
        if (!MCPClientManager.instance) {
            MCPClientManager.instance = new MCPClientManager();
        }
        return MCPClientManager.instance;
    }
    /**
     * Get or create MCP client for a server
     */
    async getClient(serverName) {
        if (!this.clients.has(serverName)) {
            const config = this.createClientConfig(serverName);
            const client = new MCPClient(config);
            await client.initialize();
            this.clients.set(serverName, client);
        }
        return this.clients.get(serverName);
    }
    /**
     * Create client configuration for a server
     */
    createClientConfig(serverName) {
        const config = defaultServerRegistry.get(serverName);
        if (config) {
            return config;
        }
        return defaultServerRegistry.createDynamicConfig(serverName);
    }
    /**
     * Call MCP server tool
     */
    async callServerTool(serverName, toolName, args = {}) {
        const client = await this.getClient(serverName);
        return client.callTool(toolName, args);
    }
    /**
     * Get all available MCP server tools
     */
    async getAllAvailableTools() {
        const jobId = `mcp-tools-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
        const result = {};
        const serverNames = [
            'code-review',
            'security-audit',
            'performance-optimization',
            'testing-strategy',
            'researcher',
            'skill-invocation',
        ];
        for (const serverName of serverNames) {
            try {
                const client = await this.getClient(serverName);
                result[serverName] = client.getAvailableTools();
            }
            catch (error) {
                frameworkLogger.log('mcp-client-manager', `failed to get tools for ${serverName}: ${error instanceof Error ? error.message : String(error)}`, 'info', { jobId });
            }
        }
        return result;
    }
    /**
     * Clear all cached clients
     */
    clearClients() {
        this.clients.clear();
    }
}
// Export singleton instance
export const mcpClientManager = MCPClientManager.getInstance();
